{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyNde72yDg8KLln4Ig6PEhGq"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Metadata\n",
        "\n",
        "Genotype. WT=control, HET=knockout Time. ZT10, ZT22, lights on is ZTO and lights off ZT12, so if ZT0 was 7am ZT10=5pm and ZT22 5am Location. 3 parts of the gut"
      ],
      "metadata": {
        "id": "dR4zVVwzh0o9"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Add google drive\n",
        "\n",
        "https://drive.google.com/drive/folders/1eyfluz240Znz6Lav67HsK7QsiNz3ypAR?usp=drive_link"
      ],
      "metadata": {
        "id": "9qoXfNJxiA24"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "dfUl3HBUh8qy",
        "outputId": "98de7431-ca91-4192-f938-72e15e9800ee"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gsutil cp /content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/*.* /content/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "slv4XXKOiHvN",
        "outputId": "1097feff-1562-44e7-ebc7-1c2d6281c6a8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/all.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/all.derep.uc...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/denovo.nonchimeras.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/derep.fasta...\n",
            "|\n",
            "==> NOTE: You are performing a sequence of gsutil operations that may\n",
            "run significantly faster if you instead use gsutil -m cp ... Please\n",
            "see the -m section under \"gsutil help options\" for further information\n",
            "about when gsutil -m can be advantageous.\n",
            "\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F10.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F10.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F10.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F10.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F11.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F11.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F11.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F11.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F12.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F12.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F12.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F12.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F13.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F13.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F13.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F13.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F14.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F14.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F14.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F14.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F15.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F15.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F15.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F15.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F16.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F16.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F16.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F16.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F17.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F17.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F17.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F17.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F18.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F18.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F18.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F18.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F19.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F19.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F19.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F19.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F1.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F1.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F1.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F1.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F20.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F20.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F20.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F20.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F21.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F21.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F21.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F21.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F22.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F22.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F22.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F22.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F23.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F23.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F23.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F23.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F24.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F24.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F24.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F24.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F25.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F25.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F25.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F25.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F26.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F26.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F26.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F26.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F27.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F27.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F27.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F27.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F28.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F28.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F28.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F28.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F29.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F29.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F29.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F29.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F2.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F2.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F2.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F2.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F30.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F30.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F30.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F30.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F31.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F31.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F31.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F31.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F3.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F3.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F3.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F3.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F4.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F4.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F4.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F4.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F5.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F5.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F5.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F5.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F6.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F6.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F6.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F6.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F7.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F7.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F7.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F7.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F8.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F8.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F8.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F8.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F9.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F9.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F9.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/F9.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/FB_ratio.csv...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/fig1b.pdf...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/fig1b.png...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/fig2.pdf...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/fig2.png...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/fig3a.pdf...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/fig3a.png...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/fig4a.pdf...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/fig4a.png...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/fig.pdf...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/fig.png...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/Hep_mouse.txt...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN10.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN10.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN10.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN10.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN11.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN11.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN11.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN11.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN12.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN12.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN12.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN12.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN13.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN13.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN13.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN13.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN14.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN14.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN14.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN14.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN15.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN15.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN15.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN15.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN16.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN16.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN16.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN16.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN18.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN18.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN18.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN18.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN19.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN19.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN19.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN19.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN1.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN1.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN1.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN1.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN20.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN20.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN20.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN20.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN21.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN21.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN21.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN21.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN22.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN22.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN22.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN22.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN23.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN23.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN23.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN23.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN24.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN24.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN24.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN24.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN25.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN25.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN25.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN25.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN2.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN2.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN2.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN2.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN3.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN3.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN3.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN3.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN4.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN4.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN4.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN4.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN5.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN5.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN5.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN5.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN6.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN6.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN6.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN6.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN7.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN7.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN7.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN7.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN8.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN8.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN8.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN8.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN9.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN9.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN9.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/IN9.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/maxEE_summary.csv...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/otus.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/otutab_log.txt...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/rare_table.txt...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/rdp_16s_v18.fa...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/sintax.txt...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/sorted.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST10.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST10.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST10.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST10.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST11.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST11.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST11.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST11.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST12.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST12.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST12.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST12.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST13.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST13.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST13.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST13.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST14.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST14.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST14.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST14.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST15.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST15.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST15.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST15.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST16.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST16.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST16.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST16.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST1.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST1.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST1.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST1.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST2.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST2.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST2.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST2.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST3.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST3.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST3.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST3.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST4.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST4.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST4.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST4.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST5.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST5.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST5.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST5.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST6.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST6.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST6.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST6.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST7.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST7.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST7.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST7.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST8.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST8.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST8.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST8.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST9.effective.fastq.gz...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST9.effective.fastq.gz.filtered.derep.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST9.effective.fastq.gz.filtered.fasta...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/ST9.effective.fastq.gz.stats...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/taxonomy.txt...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/uc_zotus.uc...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/unchime3_log.txt...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/unoise3_zotu_table.biom...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/unoise3_zotu_table.txt...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/unoise_log.txt...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/zotus.fa...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/zotus_nochime.fa...\n",
            "Copying file:///content/drive/MyDrive/BIOC0023_24_25/sequencing_data/Gut2025/zotus_nochime_nomask.fa...\n",
            "\\ [316 files][  2.3 GiB/  2.3 GiB]   79.7 MiB/s                                 \n",
            "==> NOTE: You are performing a sequence of gsutil operations that may\n",
            "run significantly faster if you instead use gsutil -m cp ... Please\n",
            "see the -m section under \"gsutil help options\" for further information\n",
            "about when gsutil -m can be advantageous.\n",
            "\n",
            "\n",
            "Operation completed over 316 objects/2.3 GiB.                                    \n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Install the packages + load"
      ],
      "metadata": {
        "id": "Uv_kBkZdiLnm"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!sudo apt-get install vsearch\n",
        "!pip install --upgrade git+https://github.com/toryn13/qiime2pandas.git"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fGVCcXU6iRFP",
        "outputId": "3df64c0e-deff-4873-f6ed-becebf812965"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "vsearch is already the newest version (2.21.1-1).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 35 not upgraded.\n",
            "Collecting git+https://github.com/toryn13/qiime2pandas.git\n",
            "  Cloning https://github.com/toryn13/qiime2pandas.git to /tmp/pip-req-build-t9hvyaum\n",
            "  Running command git clone --filter=blob:none --quiet https://github.com/toryn13/qiime2pandas.git /tmp/pip-req-build-t9hvyaum\n",
            "  Resolved https://github.com/toryn13/qiime2pandas.git to commit bdd1dd4ebc38bcd3b6ebd81e4297adc1b8a3df6b\n",
            "  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
            "  Preparing metadata (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.11/dist-packages (from qiime2pandas==0.1.0) (2.2.2)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from qiime2pandas==0.1.0) (2.0.2)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas->qiime2pandas==0.1.0) (2.9.0.post0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas->qiime2pandas==0.1.0) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas->qiime2pandas==0.1.0) (2025.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas->qiime2pandas==0.1.0) (1.17.0)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "fastq quality check"
      ],
      "metadata": {
        "id": "CKckkGP2iUJM"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "VSEARCH=\"vsearch\"\n",
        "\n",
        "for s in *.fastq.gz; do\n",
        "    echo \"Calculate maxEE statistics for sample: $s\"\n",
        "    \"$VSEARCH\" --fastq_eestats2 \"$s\" --output \"${s}.stats\"\n",
        "done"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ffLtNmWFiXYk",
        "outputId": "5c05bc30-939b-43c4-c829-dca7160e859c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Calculate maxEE statistics for sample: F10.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F11.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F12.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F13.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F14.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F15.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F16.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F17.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F18.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F19.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F1.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F20.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F21.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F22.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F23.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F24.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F25.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F26.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F27.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F28.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F29.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F2.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F30.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F31.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F3.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F4.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F5.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F6.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F7.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F8.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: F9.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN10.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN11.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN12.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN13.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN14.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN15.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN16.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN18.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN19.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN1.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN20.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN21.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN22.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN23.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN24.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN25.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN2.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN3.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN4.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN5.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN6.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN7.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN8.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: IN9.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST10.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST11.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST12.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST13.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST14.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST15.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST16.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST1.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST2.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST3.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST4.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST5.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST6.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST7.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST8.effective.fastq.gz\n",
            "Calculate maxEE statistics for sample: ST9.effective.fastq.gz\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading FASTQ file 100%\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# maxEE stats\n",
        "import qiime2pandas as q2p\n",
        "q2p.process_stats_files('/content/', 'MaxEE1')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "6HVplAa2ibmG",
        "outputId": "120a16c8-3bb9-45f4-e292-03c9f8d7f466"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "     Length  MaxEE1                          File\n",
              "0        50   50126  F10.effective.fastq.gz.stats\n",
              "1       100   49677  F10.effective.fastq.gz.stats\n",
              "2       150   48810  F10.effective.fastq.gz.stats\n",
              "3       200   47671  F10.effective.fastq.gz.stats\n",
              "4       250   46658  F10.effective.fastq.gz.stats\n",
              "..      ...     ...                           ...\n",
              "563     200   54548  F31.effective.fastq.gz.stats\n",
              "564     250   53655  F31.effective.fastq.gz.stats\n",
              "565     300   52697  F31.effective.fastq.gz.stats\n",
              "566     350   51973  F31.effective.fastq.gz.stats\n",
              "567     400   51614  F31.effective.fastq.gz.stats\n",
              "\n",
              "[568 rows x 3 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-bb6153b3-ee4d-43a8-b342-e03dbf463e92\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Length</th>\n",
              "      <th>MaxEE1</th>\n",
              "      <th>File</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>50</td>\n",
              "      <td>50126</td>\n",
              "      <td>F10.effective.fastq.gz.stats</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>100</td>\n",
              "      <td>49677</td>\n",
              "      <td>F10.effective.fastq.gz.stats</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>150</td>\n",
              "      <td>48810</td>\n",
              "      <td>F10.effective.fastq.gz.stats</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>200</td>\n",
              "      <td>47671</td>\n",
              "      <td>F10.effective.fastq.gz.stats</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>250</td>\n",
              "      <td>46658</td>\n",
              "      <td>F10.effective.fastq.gz.stats</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>563</th>\n",
              "      <td>200</td>\n",
              "      <td>54548</td>\n",
              "      <td>F31.effective.fastq.gz.stats</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>564</th>\n",
              "      <td>250</td>\n",
              "      <td>53655</td>\n",
              "      <td>F31.effective.fastq.gz.stats</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>565</th>\n",
              "      <td>300</td>\n",
              "      <td>52697</td>\n",
              "      <td>F31.effective.fastq.gz.stats</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>566</th>\n",
              "      <td>350</td>\n",
              "      <td>51973</td>\n",
              "      <td>F31.effective.fastq.gz.stats</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>567</th>\n",
              "      <td>400</td>\n",
              "      <td>51614</td>\n",
              "      <td>F31.effective.fastq.gz.stats</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>568 rows × 3 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-bb6153b3-ee4d-43a8-b342-e03dbf463e92')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-bb6153b3-ee4d-43a8-b342-e03dbf463e92 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-bb6153b3-ee4d-43a8-b342-e03dbf463e92');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-ef0b6496-98f6-4e69-9523-4c75df6f65bf\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-ef0b6496-98f6-4e69-9523-4c75df6f65bf')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-ef0b6496-98f6-4e69-9523-4c75df6f65bf button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"q2p\",\n  \"rows\": 568,\n  \"fields\": [\n    {\n      \"column\": \"Length\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 114,\n        \"min\": 50,\n        \"max\": 400,\n        \"num_unique_values\": 8,\n        \"samples\": [\n          100,\n          300,\n          50\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"MaxEE1\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 8920,\n        \"min\": 23167,\n        \"max\": 87986,\n        \"num_unique_values\": 561,\n        \"samples\": [\n          53183,\n          50798,\n          48205\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"File\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 71,\n        \"samples\": [\n          \"F20.effective.fastq.gz.stats\",\n          \"F10.effective.fastq.gz.stats\",\n          \"ST14.effective.fastq.gz.stats\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 97
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# quality filtering (VSEARCH)... trunclen 400.\n",
        "%%bash\n",
        "VSEARCH=\"vsearch\"\n",
        "for s in *.fastq.gz; do\n",
        "    echo \"Quality filtering for sample: $s\"\n",
        "    \"$VSEARCH\" --fastq_filter \"$s\" \\\n",
        "        --fastq_maxee 1 \\\n",
        "        --fastq_trunclen 400 \\\n",
        "        --fastq_maxns 0 \\\n",
        "        --fastaout \"${s%.fastq}.filtered.fasta\" \\\n",
        "        --fasta_width 0\n",
        "done"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Ksc_mjhZijkz",
        "outputId": "d4977698-79ca-41f2-a50b-8cd2d3b8c746"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Quality filtering for sample: F10.effective.fastq.gz\n",
            "Quality filtering for sample: F11.effective.fastq.gz\n",
            "Quality filtering for sample: F12.effective.fastq.gz\n",
            "Quality filtering for sample: F13.effective.fastq.gz\n",
            "Quality filtering for sample: F14.effective.fastq.gz\n",
            "Quality filtering for sample: F15.effective.fastq.gz\n",
            "Quality filtering for sample: F16.effective.fastq.gz\n",
            "Quality filtering for sample: F17.effective.fastq.gz\n",
            "Quality filtering for sample: F18.effective.fastq.gz\n",
            "Quality filtering for sample: F19.effective.fastq.gz\n",
            "Quality filtering for sample: F1.effective.fastq.gz\n",
            "Quality filtering for sample: F20.effective.fastq.gz\n",
            "Quality filtering for sample: F21.effective.fastq.gz\n",
            "Quality filtering for sample: F22.effective.fastq.gz\n",
            "Quality filtering for sample: F23.effective.fastq.gz\n",
            "Quality filtering for sample: F24.effective.fastq.gz\n",
            "Quality filtering for sample: F25.effective.fastq.gz\n",
            "Quality filtering for sample: F26.effective.fastq.gz\n",
            "Quality filtering for sample: F27.effective.fastq.gz\n",
            "Quality filtering for sample: F28.effective.fastq.gz\n",
            "Quality filtering for sample: F29.effective.fastq.gz\n",
            "Quality filtering for sample: F2.effective.fastq.gz\n",
            "Quality filtering for sample: F30.effective.fastq.gz\n",
            "Quality filtering for sample: F31.effective.fastq.gz\n",
            "Quality filtering for sample: F3.effective.fastq.gz\n",
            "Quality filtering for sample: F4.effective.fastq.gz\n",
            "Quality filtering for sample: F5.effective.fastq.gz\n",
            "Quality filtering for sample: F6.effective.fastq.gz\n",
            "Quality filtering for sample: F7.effective.fastq.gz\n",
            "Quality filtering for sample: F8.effective.fastq.gz\n",
            "Quality filtering for sample: F9.effective.fastq.gz\n",
            "Quality filtering for sample: IN10.effective.fastq.gz\n",
            "Quality filtering for sample: IN11.effective.fastq.gz\n",
            "Quality filtering for sample: IN12.effective.fastq.gz\n",
            "Quality filtering for sample: IN13.effective.fastq.gz\n",
            "Quality filtering for sample: IN14.effective.fastq.gz\n",
            "Quality filtering for sample: IN15.effective.fastq.gz\n",
            "Quality filtering for sample: IN16.effective.fastq.gz\n",
            "Quality filtering for sample: IN18.effective.fastq.gz\n",
            "Quality filtering for sample: IN19.effective.fastq.gz\n",
            "Quality filtering for sample: IN1.effective.fastq.gz\n",
            "Quality filtering for sample: IN20.effective.fastq.gz\n",
            "Quality filtering for sample: IN21.effective.fastq.gz\n",
            "Quality filtering for sample: IN22.effective.fastq.gz\n",
            "Quality filtering for sample: IN23.effective.fastq.gz\n",
            "Quality filtering for sample: IN24.effective.fastq.gz\n",
            "Quality filtering for sample: IN25.effective.fastq.gz\n",
            "Quality filtering for sample: IN2.effective.fastq.gz\n",
            "Quality filtering for sample: IN3.effective.fastq.gz\n",
            "Quality filtering for sample: IN4.effective.fastq.gz\n",
            "Quality filtering for sample: IN5.effective.fastq.gz\n",
            "Quality filtering for sample: IN6.effective.fastq.gz\n",
            "Quality filtering for sample: IN7.effective.fastq.gz\n",
            "Quality filtering for sample: IN8.effective.fastq.gz\n",
            "Quality filtering for sample: IN9.effective.fastq.gz\n",
            "Quality filtering for sample: ST10.effective.fastq.gz\n",
            "Quality filtering for sample: ST11.effective.fastq.gz\n",
            "Quality filtering for sample: ST12.effective.fastq.gz\n",
            "Quality filtering for sample: ST13.effective.fastq.gz\n",
            "Quality filtering for sample: ST14.effective.fastq.gz\n",
            "Quality filtering for sample: ST15.effective.fastq.gz\n",
            "Quality filtering for sample: ST16.effective.fastq.gz\n",
            "Quality filtering for sample: ST1.effective.fastq.gz\n",
            "Quality filtering for sample: ST2.effective.fastq.gz\n",
            "Quality filtering for sample: ST3.effective.fastq.gz\n",
            "Quality filtering for sample: ST4.effective.fastq.gz\n",
            "Quality filtering for sample: ST5.effective.fastq.gz\n",
            "Quality filtering for sample: ST6.effective.fastq.gz\n",
            "Quality filtering for sample: ST7.effective.fastq.gz\n",
            "Quality filtering for sample: ST8.effective.fastq.gz\n",
            "Quality filtering for sample: ST9.effective.fastq.gz\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "44383 sequences kept (of which 44383 truncated), 5775 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "43979 sequences kept (of which 43976 truncated), 4956 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "45905 sequences kept (of which 45905 truncated), 5127 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "36670 sequences kept (of which 36670 truncated), 4259 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "47090 sequences kept (of which 47090 truncated), 5413 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "36452 sequences kept (of which 36452 truncated), 4353 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "48660 sequences kept (of which 48660 truncated), 4989 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "52946 sequences kept (of which 52945 truncated), 6382 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "46916 sequences kept (of which 46916 truncated), 5480 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "41062 sequences kept (of which 41062 truncated), 5021 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "44061 sequences kept (of which 44061 truncated), 4786 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "43413 sequences kept (of which 43413 truncated), 5403 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "48924 sequences kept (of which 48924 truncated), 5914 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "50764 sequences kept (of which 50764 truncated), 5713 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "45322 sequences kept (of which 45322 truncated), 4896 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "48525 sequences kept (of which 48525 truncated), 5666 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "51731 sequences kept (of which 51731 truncated), 5857 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "48281 sequences kept (of which 48281 truncated), 5776 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "52033 sequences kept (of which 52031 truncated), 6440 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "52796 sequences kept (of which 52793 truncated), 6436 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "50794 sequences kept (of which 50794 truncated), 6252 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "48691 sequences kept (of which 48691 truncated), 5235 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "47254 sequences kept (of which 47253 truncated), 5857 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "51602 sequences kept (of which 51602 truncated), 6447 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "39577 sequences kept (of which 39577 truncated), 4362 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "40042 sequences kept (of which 40042 truncated), 4546 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "37268 sequences kept (of which 37268 truncated), 4940 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "43324 sequences kept (of which 43323 truncated), 5087 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "45350 sequences kept (of which 45350 truncated), 5050 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "41437 sequences kept (of which 41437 truncated), 4968 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "48600 sequences kept (of which 48600 truncated), 5696 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "51431 sequences kept (of which 51431 truncated), 5259 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "47450 sequences kept (of which 47450 truncated), 5587 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "26255 sequences kept (of which 26255 truncated), 3418 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "61606 sequences kept (of which 61606 truncated), 7122 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "46409 sequences kept (of which 46409 truncated), 6622 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "45724 sequences kept (of which 45724 truncated), 5323 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "52732 sequences kept (of which 52732 truncated), 5970 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "55686 sequences kept (of which 55686 truncated), 7187 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "52929 sequences kept (of which 52929 truncated), 5380 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "53031 sequences kept (of which 53030 truncated), 6546 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "33143 sequences kept (of which 33143 truncated), 3523 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "42039 sequences kept (of which 42039 truncated), 6200 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "52767 sequences kept (of which 52767 truncated), 6318 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "53086 sequences kept (of which 53086 truncated), 5918 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "56032 sequences kept (of which 56031 truncated), 9363 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "49283 sequences kept (of which 49283 truncated), 5703 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "52854 sequences kept (of which 52854 truncated), 5633 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "50330 sequences kept (of which 50330 truncated), 6485 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "45760 sequences kept (of which 45760 truncated), 5442 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "49397 sequences kept (of which 49396 truncated), 5280 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "55776 sequences kept (of which 55776 truncated), 6318 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "56301 sequences kept (of which 56301 truncated), 5991 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "63587 sequences kept (of which 63586 truncated), 7839 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "52666 sequences kept (of which 52666 truncated), 5837 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "76164 sequences kept (of which 76164 truncated), 11878 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "53158 sequences kept (of which 53158 truncated), 6061 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "53910 sequences kept (of which 53910 truncated), 6241 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "57135 sequences kept (of which 57135 truncated), 6309 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "50685 sequences kept (of which 50683 truncated), 6953 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "53276 sequences kept (of which 53276 truncated), 5894 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "28820 sequences kept (of which 28807 truncated), 16952 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "50186 sequences kept (of which 50186 truncated), 10236 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "23166 sequences kept (of which 23166 truncated), 2984 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "56368 sequences kept (of which 56368 truncated), 6414 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "53518 sequences kept (of which 53518 truncated), 6418 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "32079 sequences kept (of which 32079 truncated), 3396 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "55922 sequences kept (of which 55922 truncated), 6210 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "55222 sequences kept (of which 55221 truncated), 6668 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "49770 sequences kept (of which 49769 truncated), 6015 sequences discarded.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "55920 sequences kept (of which 55918 truncated), 6236 sequences discarded.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Sample dereplication"
      ],
      "metadata": {
        "id": "6rKrOyzSivrQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "VSEARCH=\"vsearch\"\n",
        "\n",
        "for s in *.filtered.fasta; do\n",
        "    echo \"Dereplicate at sample level and relabel with sample.n for: $s\"\n",
        "    \"$VSEARCH\" --derep_fulllength \"$s\" \\\n",
        "        --strand plus \\\n",
        "        --output \"${s%.fasta}.derep.fasta\" \\\n",
        "        --sizeout \\\n",
        "        --relabel \"${s%.filtered.fasta}.\" \\\n",
        "        --fasta_width 0\n",
        "done"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SkR31hPciunD",
        "outputId": "1070d767-abf7-4a44-e87c-e653fdf78665"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Dereplicate at sample level and relabel with sample.n for: F10.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F11.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F12.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F13.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F14.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F15.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F16.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F17.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F18.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F19.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F1.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F20.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F21.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F22.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F23.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F24.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F25.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F26.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F27.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F28.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F29.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F2.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F30.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F31.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F3.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F4.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F5.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F6.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F7.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F8.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: F9.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN10.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN11.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN12.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN13.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN14.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN15.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN16.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN18.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN19.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN1.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN20.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN21.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN22.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN23.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN24.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN25.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN2.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN3.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN4.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN5.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN6.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN7.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN8.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: IN9.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST10.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST11.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST12.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST13.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST14.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST15.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST16.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST1.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST2.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST3.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST4.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST5.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST6.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST7.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST8.effective.fastq.gz.filtered.fasta\n",
            "Dereplicate at sample level and relabel with sample.n for: ST9.effective.fastq.gz.filtered.fasta\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F10.effective.fastq.gz.filtered.fasta 100%\n",
            "17753200 nt in 44383 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "6217 unique sequences, avg cluster 7.1, median 1, max 7199\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F11.effective.fastq.gz.filtered.fasta 100%\n",
            "17591600 nt in 43979 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7231 unique sequences, avg cluster 6.1, median 1, max 5092\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F12.effective.fastq.gz.filtered.fasta 100%\n",
            "18362000 nt in 45905 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7220 unique sequences, avg cluster 6.4, median 1, max 5326\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F13.effective.fastq.gz.filtered.fasta 100%\n",
            "14668000 nt in 36670 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7820 unique sequences, avg cluster 4.7, median 1, max 1477\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F14.effective.fastq.gz.filtered.fasta 100%\n",
            "18836000 nt in 47090 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "9815 unique sequences, avg cluster 4.8, median 1, max 1697\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F15.effective.fastq.gz.filtered.fasta 100%\n",
            "14580800 nt in 36452 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7515 unique sequences, avg cluster 4.9, median 1, max 1968\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F16.effective.fastq.gz.filtered.fasta 100%\n",
            "19464000 nt in 48660 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8961 unique sequences, avg cluster 5.4, median 1, max 3097\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F17.effective.fastq.gz.filtered.fasta 100%\n",
            "21178400 nt in 52946 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "9132 unique sequences, avg cluster 5.8, median 1, max 5150\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F18.effective.fastq.gz.filtered.fasta 100%\n",
            "18766400 nt in 46916 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8161 unique sequences, avg cluster 5.7, median 1, max 2684\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F19.effective.fastq.gz.filtered.fasta 100%\n",
            "16424800 nt in 41062 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7589 unique sequences, avg cluster 5.4, median 1, max 2641\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F1.effective.fastq.gz.filtered.fasta 100%\n",
            "17624400 nt in 44061 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7793 unique sequences, avg cluster 5.7, median 1, max 2485\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F20.effective.fastq.gz.filtered.fasta 100%\n",
            "17365200 nt in 43413 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8349 unique sequences, avg cluster 5.2, median 1, max 2121\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F21.effective.fastq.gz.filtered.fasta 100%\n",
            "19569600 nt in 48924 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8065 unique sequences, avg cluster 6.1, median 1, max 5374\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F22.effective.fastq.gz.filtered.fasta 100%\n",
            "20305600 nt in 50764 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8215 unique sequences, avg cluster 6.2, median 1, max 4836\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F23.effective.fastq.gz.filtered.fasta 100%\n",
            "18128800 nt in 45322 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7293 unique sequences, avg cluster 6.2, median 1, max 7145\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F24.effective.fastq.gz.filtered.fasta 100%\n",
            "19410000 nt in 48525 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7294 unique sequences, avg cluster 6.7, median 1, max 3233\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F25.effective.fastq.gz.filtered.fasta 100%\n",
            "20692400 nt in 51731 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7850 unique sequences, avg cluster 6.6, median 1, max 3457\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F26.effective.fastq.gz.filtered.fasta 100%\n",
            "19312400 nt in 48281 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7160 unique sequences, avg cluster 6.7, median 1, max 4264\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F27.effective.fastq.gz.filtered.fasta 100%\n",
            "20813200 nt in 52033 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8008 unique sequences, avg cluster 6.5, median 1, max 2505\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F28.effective.fastq.gz.filtered.fasta 100%\n",
            "21118400 nt in 52796 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7315 unique sequences, avg cluster 7.2, median 1, max 4748\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F29.effective.fastq.gz.filtered.fasta 100%\n",
            "20317600 nt in 50794 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7831 unique sequences, avg cluster 6.5, median 1, max 2863\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F2.effective.fastq.gz.filtered.fasta 100%\n",
            "19476400 nt in 48691 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7756 unique sequences, avg cluster 6.3, median 1, max 7073\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F30.effective.fastq.gz.filtered.fasta 100%\n",
            "18901600 nt in 47254 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "6937 unique sequences, avg cluster 6.8, median 1, max 3311\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F31.effective.fastq.gz.filtered.fasta 100%\n",
            "20640800 nt in 51602 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7214 unique sequences, avg cluster 7.2, median 1, max 3865\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F3.effective.fastq.gz.filtered.fasta 100%\n",
            "15830800 nt in 39577 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8282 unique sequences, avg cluster 4.8, median 1, max 2118\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F4.effective.fastq.gz.filtered.fasta 100%\n",
            "16016800 nt in 40042 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8510 unique sequences, avg cluster 4.7, median 1, max 1892\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F5.effective.fastq.gz.filtered.fasta 100%\n",
            "14907200 nt in 37268 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "6677 unique sequences, avg cluster 5.6, median 1, max 4311\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F6.effective.fastq.gz.filtered.fasta 100%\n",
            "17329600 nt in 43324 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7176 unique sequences, avg cluster 6.0, median 1, max 3579\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F7.effective.fastq.gz.filtered.fasta 100%\n",
            "18140000 nt in 45350 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8225 unique sequences, avg cluster 5.5, median 1, max 4279\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F8.effective.fastq.gz.filtered.fasta 100%\n",
            "16574800 nt in 41437 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "9170 unique sequences, avg cluster 4.5, median 1, max 1461\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file F9.effective.fastq.gz.filtered.fasta 100%\n",
            "19440000 nt in 48600 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8044 unique sequences, avg cluster 6.0, median 1, max 7073\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN10.effective.fastq.gz.filtered.fasta 100%\n",
            "20572400 nt in 51431 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7975 unique sequences, avg cluster 6.4, median 1, max 8980\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN11.effective.fastq.gz.filtered.fasta 100%\n",
            "18980000 nt in 47450 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7646 unique sequences, avg cluster 6.2, median 1, max 4481\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN12.effective.fastq.gz.filtered.fasta 100%\n",
            "10502000 nt in 26255 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "3449 unique sequences, avg cluster 7.6, median 1, max 11450\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN13.effective.fastq.gz.filtered.fasta 100%\n",
            "24642400 nt in 61606 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "9856 unique sequences, avg cluster 6.3, median 1, max 9009\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN14.effective.fastq.gz.filtered.fasta 100%\n",
            "18563600 nt in 46409 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "6409 unique sequences, avg cluster 7.2, median 1, max 7388\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN15.effective.fastq.gz.filtered.fasta 100%\n",
            "18289600 nt in 45724 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7661 unique sequences, avg cluster 6.0, median 1, max 6222\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN16.effective.fastq.gz.filtered.fasta 100%\n",
            "21092800 nt in 52732 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8605 unique sequences, avg cluster 6.1, median 1, max 5038\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN18.effective.fastq.gz.filtered.fasta 100%\n",
            "22274400 nt in 55686 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "6255 unique sequences, avg cluster 8.9, median 1, max 13635\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN19.effective.fastq.gz.filtered.fasta 100%\n",
            "21171600 nt in 52929 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8223 unique sequences, avg cluster 6.4, median 1, max 8047\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN1.effective.fastq.gz.filtered.fasta 100%\n",
            "21212400 nt in 53031 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "5237 unique sequences, avg cluster 10.1, median 1, max 11206\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN20.effective.fastq.gz.filtered.fasta 100%\n",
            "13257200 nt in 33143 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "3882 unique sequences, avg cluster 8.5, median 1, max 11986\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN21.effective.fastq.gz.filtered.fasta 100%\n",
            "16815600 nt in 42039 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "4217 unique sequences, avg cluster 10.0, median 1, max 16714\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN22.effective.fastq.gz.filtered.fasta 100%\n",
            "21106800 nt in 52767 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7551 unique sequences, avg cluster 7.0, median 1, max 4042\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN23.effective.fastq.gz.filtered.fasta 100%\n",
            "21234400 nt in 53086 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "6913 unique sequences, avg cluster 7.7, median 1, max 10419\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN24.effective.fastq.gz.filtered.fasta 100%\n",
            "22412800 nt in 56032 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "6576 unique sequences, avg cluster 8.5, median 1, max 15410\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN25.effective.fastq.gz.filtered.fasta 100%\n",
            "19713200 nt in 49283 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "6991 unique sequences, avg cluster 7.0, median 1, max 12006\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN2.effective.fastq.gz.filtered.fasta 100%\n",
            "21141600 nt in 52854 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "4443 unique sequences, avg cluster 11.9, median 1, max 21694\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN3.effective.fastq.gz.filtered.fasta 100%\n",
            "20132000 nt in 50330 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "4966 unique sequences, avg cluster 10.1, median 1, max 12181\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN4.effective.fastq.gz.filtered.fasta 100%\n",
            "18304000 nt in 45760 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8959 unique sequences, avg cluster 5.1, median 1, max 3842\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN5.effective.fastq.gz.filtered.fasta 100%\n",
            "19758800 nt in 49397 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7905 unique sequences, avg cluster 6.2, median 1, max 8551\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN6.effective.fastq.gz.filtered.fasta 100%\n",
            "22310400 nt in 55776 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8079 unique sequences, avg cluster 6.9, median 1, max 6919\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN7.effective.fastq.gz.filtered.fasta 100%\n",
            "22520400 nt in 56301 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7856 unique sequences, avg cluster 7.2, median 1, max 7765\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN8.effective.fastq.gz.filtered.fasta 100%\n",
            "25434800 nt in 63587 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8454 unique sequences, avg cluster 7.5, median 1, max 12014\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file IN9.effective.fastq.gz.filtered.fasta 100%\n",
            "21066400 nt in 52666 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8102 unique sequences, avg cluster 6.5, median 1, max 7803\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST10.effective.fastq.gz.filtered.fasta 100%\n",
            "30465600 nt in 76164 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "4617 unique sequences, avg cluster 16.5, median 1, max 51233\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST11.effective.fastq.gz.filtered.fasta 100%\n",
            "21263200 nt in 53158 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8190 unique sequences, avg cluster 6.5, median 1, max 6754\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST12.effective.fastq.gz.filtered.fasta 100%\n",
            "21564000 nt in 53910 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8509 unique sequences, avg cluster 6.3, median 1, max 6347\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST13.effective.fastq.gz.filtered.fasta 100%\n",
            "22854000 nt in 57135 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "6728 unique sequences, avg cluster 8.5, median 1, max 11448\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST14.effective.fastq.gz.filtered.fasta 100%\n",
            "20274000 nt in 50685 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "7269 unique sequences, avg cluster 7.0, median 1, max 7769\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST15.effective.fastq.gz.filtered.fasta 100%\n",
            "21310400 nt in 53276 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8021 unique sequences, avg cluster 6.6, median 1, max 4399\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST16.effective.fastq.gz.filtered.fasta 100%\n",
            "11528000 nt in 28820 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "2945 unique sequences, avg cluster 9.8, median 1, max 13699\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST1.effective.fastq.gz.filtered.fasta 100%\n",
            "20074400 nt in 50186 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "4928 unique sequences, avg cluster 10.2, median 1, max 9531\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST2.effective.fastq.gz.filtered.fasta 100%\n",
            "9266400 nt in 23166 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "1714 unique sequences, avg cluster 13.5, median 1, max 17364\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST3.effective.fastq.gz.filtered.fasta 100%\n",
            "22547200 nt in 56368 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "4670 unique sequences, avg cluster 12.1, median 1, max 30341\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST4.effective.fastq.gz.filtered.fasta 100%\n",
            "21407200 nt in 53518 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "9266 unique sequences, avg cluster 5.8, median 1, max 3361\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST5.effective.fastq.gz.filtered.fasta 100%\n",
            "12831600 nt in 32079 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "6948 unique sequences, avg cluster 4.6, median 1, max 2251\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST6.effective.fastq.gz.filtered.fasta 100%\n",
            "22368800 nt in 55922 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "4697 unique sequences, avg cluster 11.9, median 1, max 20448\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST7.effective.fastq.gz.filtered.fasta 100%\n",
            "22088800 nt in 55222 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "8392 unique sequences, avg cluster 6.6, median 1, max 12741\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST8.effective.fastq.gz.filtered.fasta 100%\n",
            "19908000 nt in 49770 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "10219 unique sequences, avg cluster 4.9, median 1, max 1177\n",
            "Writing FASTA output file 100%\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file ST9.effective.fastq.gz.filtered.fasta 100%\n",
            "22368000 nt in 55920 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "6312 unique sequences, avg cluster 8.9, median 1, max 12773\n",
            "Writing FASTA output file 100%\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# sample dereplication results report\n",
        "%%bash\n",
        "\n",
        "cat *.derep.fasta > all.derep.fasta\n",
        "echo \"Sum of unique sequences in each sample: $(grep -c \"^>\" all.derep.fasta)\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lFvX08o-i1EJ",
        "outputId": "bb0521c9-01dc-48da-cb88-d78eb2109b5e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Sum of unique sequences in each sample: 512460\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# number of unique sequences?\n",
        "%%bash\n",
        "VSEARCH=\"vsearch\"\n",
        "\n",
        "\"$VSEARCH\" --derep_fulllength all.derep.fasta \\\n",
        "    --sizein \\\n",
        "    --sizeout \\\n",
        "    --fasta_width 0 \\\n",
        "    --uc all.derep.uc \\\n",
        "    --output derep.fasta\n",
        "echo \"Unique sequences: $(grep -c \"^>\" derep.fasta)\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ikYaZOKLi-Y8",
        "outputId": "79d8a415-906f-4460-c8ad-536292cae717"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Unique sequences: 256628\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Dereplicating file all.derep.fasta 100%\n",
            "204984000 nt in 512460 seqs, min 400, max 400, avg 400\n",
            "Sorting 100%\n",
            "256628 unique sequences, avg cluster 13.4, median 1, max 208892\n",
            "Writing FASTA output file 100%\n",
            "Writing uc file, first part 100%\n",
            "Writing uc file, second part 100%\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Clustering ASVs. VSEARCH"
      ],
      "metadata": {
        "id": "DgPQWzWUjHSb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# use UNOISE\n",
        "%%bash\n",
        "VSEARCH=\"vsearch\"\n",
        "\n",
        "\"$VSEARCH\" --cluster_unoise all.derep.fasta \\\n",
        "           --centroids zotus.fa \\\n",
        "            --uc uc_zotus.uc \\\n",
        "            --log unoise_log.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "evR1zd9PjF_U",
        "outputId": "19bfcda7-ba95-4667-e4d4-768c2f976fc0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading file all.derep.fasta 100%\n",
            "9232800 nt in 23082 seqs, min 400, max 400, avg 400\n",
            "minsize 8: 489378 sequences discarded.\n",
            "Masking 100%\n",
            "Sorting by abundance 100%\n",
            "Counting k-mers 100%\n",
            "Clustering 100%\n",
            "Sorting clusters 100%\n",
            "Writing clusters 100%\n",
            "Clusters: 1555 Size min 1, max 1407, avg 14.8\n",
            "Singletons: 906, 3.9% of seqs, 58.3% of clusters\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Singleton removal"
      ],
      "metadata": {
        "id": "QYXb5lFujUex"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# singleton removal results report\n",
        "# redundant here... consider rearrangement\n",
        "%%bash\n",
        "\n",
        "VSEARCH=\"vsearch\"\n",
        "THREADS=2\n",
        "\n",
        "$VSEARCH --sortbysize zotus.fa \\\n",
        "    --threads $THREADS \\\n",
        "    --minsize 2 \\\n",
        "    --fasta_width 0 \\\n",
        "   --output sorted.fasta\n",
        "\n",
        "echo \"Non-singleton clusters: $(grep -c \"^>\" sorted.fasta)\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KYd-tFsHjTgD",
        "outputId": "094ac725-c0f2-49c5-be5e-07dd1bdfb3ed"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Non-singleton clusters: 1555\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING: The sortbysize command does not support multithreading.\n",
            "Only 1 thread used.\n",
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading file zotus.fa 100%\n",
            "622000 nt in 1555 seqs, min 400, max 400, avg 400\n",
            "Getting sizes 100%\n",
            "Sorting 100%\n",
            "Median abundance: 34\n",
            "Writing output 100%\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Chimera removal"
      ],
      "metadata": {
        "id": "ZU33pTtIjgqt"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# chimera removal results report\n",
        "%%bash\n",
        "\n",
        "VSEARCH=\"vsearch\"\n",
        "\n",
        "$VSEARCH --uchime_denovo sorted.fasta \\\n",
        "    --fasta_width 0 \\\n",
        "    --nonchimeras denovo.nonchimeras.fasta\n",
        "\n",
        "echo \"Unique sequences after de novo chimera detection: $(grep -c \"^>\" denovo.nonchimeras.fasta)\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "R_p7kukzjh1-",
        "outputId": "c909ed6b-ebbe-429c-b412-2e18fd691e12"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Unique sequences after de novo chimera detection: 1469\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading file sorted.fasta 100%\n",
            "622000 nt in 1555 seqs, min 400, max 400, avg 400\n",
            "Masking 100%\n",
            "Sorting by abundance 100%\n",
            "Counting k-mers 100%\n",
            "Detecting chimeras 100%\n",
            "Found 69 (4.4%) chimeras, 1469 (94.5%) non-chimeras,\n",
            "and 17 (1.1%) borderline sequences in 1555 unique sequences.\n",
            "Taking abundance information into account, this corresponds to\n",
            "3734 (0.9%) chimeras, 424024 (99.0%) non-chimeras,\n",
            "and 464 (0.1%) borderline sequences in 428222 total sequences.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# ASV number report\n",
        "%%bash\n",
        "\n",
        "VSEARCH=\"vsearch\"\n",
        "\n",
        "$VSEARCH --fastx_filter denovo.nonchimeras.fasta \\\n",
        "    --fasta_width 0 \\\n",
        "    --relabel OTU_ \\\n",
        "    --fastaout otus.fasta\n",
        "\n",
        "echo \"Number of ASVs: $(grep -c \"^>\" otus.fasta)\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "u3LUZWifj5ds",
        "outputId": "ea6bc4b2-4234-4d66-9bc4-904d7f93cfcc"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of ASVs: 1469\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading input file 100%\n",
            "1469 sequences kept (of which 0 truncated), 0 sequences discarded.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "de novo non-chimera"
      ],
      "metadata": {
        "id": "7Ky8jqRlkF0F"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "vsearch --cluster_unoise all.derep.fasta --centroids zotus.fa --uc uc_zotus.uc --log unoise_log.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "M3BLuNWwkFMz",
        "outputId": "277ab6d3-1dc6-4e29-a962-f0fb2f27dd98"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading file all.derep.fasta 100%\n",
            "9232800 nt in 23082 seqs, min 400, max 400, avg 400\n",
            "minsize 8: 489378 sequences discarded.\n",
            "Masking 100%\n",
            "Sorting by abundance 100%\n",
            "Counting k-mers 100%\n",
            "Clustering 100%\n",
            "Sorting clusters 100%\n",
            "Writing clusters 100%\n",
            "Clusters: 1555 Size min 1, max 1407, avg 14.8\n",
            "Singletons: 906, 3.9% of seqs, 58.3% of clusters\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Normalize to uppercase\n",
        "awk 'BEGIN{FS=\" \"}{if(!/>/){print toupper($0)}else{print $1}}' zotus_nochime.fa > zotus_nochime_nomask.fa\n",
        "\n",
        "# Map reads back to centroids\n",
        "vsearch --usearch_global all.derep.fasta --db zotus_nochime_nomask.fa --id 0.97 --otutabout unoise3_zotu_table.txt --biomout unoise3_zotu_table.biom  --log otutab_log.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "csYBLFbTkOGW",
        "outputId": "6963ffe3-e259-444e-b255-79a354c66f8c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "vsearch v2.21.1_linux_x86_64, 12.7GB RAM, 2 cores\n",
            "https://github.com/torognes/vsearch\n",
            "\n",
            "Reading file zotus_nochime_nomask.fa 100%\n",
            "601600 nt in 1504 seqs, min 400, max 400, avg 400\n",
            "Masking 100%\n",
            "Counting k-mers 100%\n",
            "Creating k-mer index 100%\n",
            "Searching 100%\n",
            "Matching unique query sequences: 505351 of 512460 (98.61%)\n",
            "Writing OTU table (biom 1.0) 100%\n",
            "Writing OTU table (classic) 100%\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Taxonomy assignment"
      ],
      "metadata": {
        "id": "s4SMwfu-kTJD"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "sintax from RDP database."
      ],
      "metadata": {
        "id": "IhH2Y6laklzY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://github.com/rcedgar/usearch12/releases/download/v12.0-beta1/usearch_linux_x86_12.0-beta"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pKySJFgakXWF",
        "outputId": "ce148e0a-a299-4a95-ad4f-70173d176ca9"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-06-04 14:18:34--  https://github.com/rcedgar/usearch12/releases/download/v12.0-beta1/usearch_linux_x86_12.0-beta\n",
            "Resolving github.com (github.com)... 140.82.121.3\n",
            "Connecting to github.com (github.com)|140.82.121.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/758594776/a3d3410c-6a1d-4855-82c9-03634617ed16?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250604%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250604T141834Z&X-Amz-Expires=300&X-Amz-Signature=7021a2773a6a6319572cba759e6ff091704477a52bb1fed4cf5938935b79ab8e&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dusearch_linux_x86_12.0-beta&response-content-type=application%2Foctet-stream [following]\n",
            "--2025-06-04 14:18:34--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/758594776/a3d3410c-6a1d-4855-82c9-03634617ed16?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250604%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250604T141834Z&X-Amz-Expires=300&X-Amz-Signature=7021a2773a6a6319572cba759e6ff091704477a52bb1fed4cf5938935b79ab8e&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dusearch_linux_x86_12.0-beta&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 2306888 (2.2M) [application/octet-stream]\n",
            "Saving to: ‘usearch_linux_x86_12.0-beta’\n",
            "\n",
            "usearch_linux_x86_1 100%[===================>]   2.20M  --.-KB/s    in 0.04s   \n",
            "\n",
            "2025-06-04 14:18:35 (49.0 MB/s) - ‘usearch_linux_x86_12.0-beta’ saved [2306888/2306888]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!mv /content/usearch_linux_x86_12.0-beta /content/usearch12\n",
        "!chmod +x /content/usearch12"
      ],
      "metadata": {
        "id": "kHOh5rG_kcjl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!curl -o rdp_16s_v18.fa.gz https://drive5.com/sintax/rdp_16s_v18.fa.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SuAk4ufQkc7z",
        "outputId": "3cc5d7f5-1898-4cb8-8a33-b622d92bc6e2"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100 6685k  100 6685k    0     0  4037k      0  0:00:01  0:00:01 --:--:-- 4039k\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# !gunzip /content/rdp_16s_v18.fa.gz"
      ],
      "metadata": {
        "id": "uoE_RcwHkiMR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# cutoff 0.8\n",
        "!./usearch12 -sintax /content/zotus_nochime.fa -db rdp_16s_v18.fa -strand both \\\n",
        "  -tabbedout sintax.txt -sintax_cutoff 0.8"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "B8ru-iOhklF8",
        "outputId": "f5e2485c-842e-452d-b346-57676c426b25"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "usearch v12.0 [b1d935b], 13.3Gb RAM, 2 cores\n",
            "(C) Copyright 2013-24 Robert C. Edgar.\n",
            "https://drive5.com/usearch\n",
            "\n",
            "sintax zotus_nochime.fa\n",
            "00:00  70Mb ...... Loading seqs.\n",
            "00:01  37Mb 100.0% Masking (fastnucleo)\n",
            "00:02  38Mb 100.0% Udb seqs\n",
            "00:02  38Mb 100.0% Alloc rows\n",
            "00:05 152Mb 100.0% Build udb index\n",
            "00:05 151Mb ...... Starting 2 threads\n",
            "00:34 158Mb 100.0% Classifying 724 / 6016 predicted genus (12.0%)\n",
            "00:34 158Mb ...... Finished\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "table=pd.read_table('/content/unoise3_zotu_table.txt', sep='\\t', index_col=0)\n",
        "\n",
        "table\n",
        "\n",
        "table.sum().sum()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5XGPYlJ_k4iy",
        "outputId": "36de157d-7902-490f-f770-46eb70b6158b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "np.int64(3425041)"
            ]
          },
          "metadata": {},
          "execution_count": 113
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Rarefaction"
      ],
      "metadata": {
        "id": "LjG5-9K6lCxM"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install scikit-bio"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3e-bwBJIlIk5",
        "outputId": "99853eec-c8fa-4d3d-f947-3ec7e88a6725"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: scikit-bio in /usr/local/lib/python3.11/dist-packages (0.6.3)\n",
            "Requirement already satisfied: requests>=2.20.0 in /usr/local/lib/python3.11/dist-packages (from scikit-bio) (2.32.3)\n",
            "Requirement already satisfied: decorator>=3.4.2 in /usr/local/lib/python3.11/dist-packages (from scikit-bio) (4.4.2)\n",
            "Requirement already satisfied: natsort>=4.0.3 in /usr/local/lib/python3.11/dist-packages (from scikit-bio) (8.4.0)\n",
            "Requirement already satisfied: numpy>=1.17.0 in /usr/local/lib/python3.11/dist-packages (from scikit-bio) (2.0.2)\n",
            "Requirement already satisfied: pandas>=1.5.0 in /usr/local/lib/python3.11/dist-packages (from scikit-bio) (2.2.2)\n",
            "Requirement already satisfied: scipy>=1.9.0 in /usr/local/lib/python3.11/dist-packages (from scikit-bio) (1.15.3)\n",
            "Requirement already satisfied: h5py>=3.6.0 in /usr/local/lib/python3.11/dist-packages (from scikit-bio) (3.13.0)\n",
            "Requirement already satisfied: biom-format>=2.1.16 in /usr/local/lib/python3.11/dist-packages (from scikit-bio) (2.1.16)\n",
            "Requirement already satisfied: statsmodels>=0.14.0 in /usr/local/lib/python3.11/dist-packages (from scikit-bio) (0.14.4)\n",
            "Requirement already satisfied: patsy>=0.5.0 in /usr/local/lib/python3.11/dist-packages (from scikit-bio) (1.0.1)\n",
            "Requirement already satisfied: click in /usr/local/lib/python3.11/dist-packages (from biom-format>=2.1.16->scikit-bio) (8.2.1)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas>=1.5.0->scikit-bio) (2.9.0.post0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas>=1.5.0->scikit-bio) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas>=1.5.0->scikit-bio) (2025.2)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests>=2.20.0->scikit-bio) (3.4.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests>=2.20.0->scikit-bio) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests>=2.20.0->scikit-bio) (2.4.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests>=2.20.0->scikit-bio) (2025.4.26)\n",
            "Requirement already satisfied: packaging>=21.3 in /usr/local/lib/python3.11/dist-packages (from statsmodels>=0.14.0->scikit-bio) (24.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas>=1.5.0->scikit-bio) (1.17.0)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "q2p.rarefaction_curve(table, max_depth=23151, num_iterations=10, seed=1976)\n",
        "plt.savefig('fig1b.png')\n",
        "plt.savefig('fig1b.pdf')"
      ],
      "metadata": {
        "id": "hu6B6dTWlE_n"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from skbio.stats import subsample_counts\n",
        "from skbio import Table\n",
        "\n",
        "table = Table(table.values, table.index, table.columns)\n",
        "rare_table = table.subsample(23151, seed=2025, axis='sample', with_replacement=False)\n",
        "rare_table.to_dataframe().to_csv('rare_table.txt', sep='\\t')"
      ],
      "metadata": {
        "id": "NYKqmpx1lO0E"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "metadata=pd.read_table('/content/Hep_mouse.txt', index_col=0)\n",
        "metadata"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 455
        },
        "id": "n5hMt2b5Soov",
        "outputId": "02230d2a-3917-44ce-fe69-a65d43206daf"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                Novogene ID Sample Name in Report Genotype  Time  Location  \\\n",
              "Sample_id                                                                    \n",
              "F1         FKDN250100713-1A                FHET10      HET    10   Faecal    \n",
              "F2         FKDN250100714-1A                 FWT10       WT    10   Faecal    \n",
              "F3         FKDN250100715-1A                 FWT22       WT    22   Faecal    \n",
              "F4         FKDN250100716-1A                FHET10      HET    10   Faecal    \n",
              "F5         FKDN250100717-1A                FHET10       WT    10   Faecal    \n",
              "...                     ...                   ...      ...   ...       ...   \n",
              "F28        FKDN250107294-1A                 FWT22      HET    22   Faecal    \n",
              "F29        FKDN250107295-1A                FHET22      HET    22   Faecal    \n",
              "F30        FKDN250107296-1A                 FWT10       WT    10   Faecal    \n",
              "F31        FKDN250107297-1A                 FWT10       WT    10   Faecal    \n",
              "ST16       FKDN250109364-1A                  WT10       WT    10  Duodenum   \n",
              "\n",
              "            Group  \n",
              "Sample_id          \n",
              "F1         HET_10  \n",
              "F2          WT_10  \n",
              "F3          WT_22  \n",
              "F4         HET_10  \n",
              "F5          WT_10  \n",
              "...           ...  \n",
              "F28        HET_22  \n",
              "F29        HET_22  \n",
              "F30         WT_10  \n",
              "F31         WT_10  \n",
              "ST16        WT_10  \n",
              "\n",
              "[71 rows x 6 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-06f9de47-ff8a-4f44-902d-469b1806dda5\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Novogene ID</th>\n",
              "      <th>Sample Name in Report</th>\n",
              "      <th>Genotype</th>\n",
              "      <th>Time</th>\n",
              "      <th>Location</th>\n",
              "      <th>Group</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Sample_id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>F1</th>\n",
              "      <td>FKDN250100713-1A</td>\n",
              "      <td>FHET10</td>\n",
              "      <td>HET</td>\n",
              "      <td>10</td>\n",
              "      <td>Faecal</td>\n",
              "      <td>HET_10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F2</th>\n",
              "      <td>FKDN250100714-1A</td>\n",
              "      <td>FWT10</td>\n",
              "      <td>WT</td>\n",
              "      <td>10</td>\n",
              "      <td>Faecal</td>\n",
              "      <td>WT_10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F3</th>\n",
              "      <td>FKDN250100715-1A</td>\n",
              "      <td>FWT22</td>\n",
              "      <td>WT</td>\n",
              "      <td>22</td>\n",
              "      <td>Faecal</td>\n",
              "      <td>WT_22</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F4</th>\n",
              "      <td>FKDN250100716-1A</td>\n",
              "      <td>FHET10</td>\n",
              "      <td>HET</td>\n",
              "      <td>10</td>\n",
              "      <td>Faecal</td>\n",
              "      <td>HET_10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F5</th>\n",
              "      <td>FKDN250100717-1A</td>\n",
              "      <td>FHET10</td>\n",
              "      <td>WT</td>\n",
              "      <td>10</td>\n",
              "      <td>Faecal</td>\n",
              "      <td>WT_10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F28</th>\n",
              "      <td>FKDN250107294-1A</td>\n",
              "      <td>FWT22</td>\n",
              "      <td>HET</td>\n",
              "      <td>22</td>\n",
              "      <td>Faecal</td>\n",
              "      <td>HET_22</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F29</th>\n",
              "      <td>FKDN250107295-1A</td>\n",
              "      <td>FHET22</td>\n",
              "      <td>HET</td>\n",
              "      <td>22</td>\n",
              "      <td>Faecal</td>\n",
              "      <td>HET_22</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F30</th>\n",
              "      <td>FKDN250107296-1A</td>\n",
              "      <td>FWT10</td>\n",
              "      <td>WT</td>\n",
              "      <td>10</td>\n",
              "      <td>Faecal</td>\n",
              "      <td>WT_10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F31</th>\n",
              "      <td>FKDN250107297-1A</td>\n",
              "      <td>FWT10</td>\n",
              "      <td>WT</td>\n",
              "      <td>10</td>\n",
              "      <td>Faecal</td>\n",
              "      <td>WT_10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ST16</th>\n",
              "      <td>FKDN250109364-1A</td>\n",
              "      <td>WT10</td>\n",
              "      <td>WT</td>\n",
              "      <td>10</td>\n",
              "      <td>Duodenum</td>\n",
              "      <td>WT_10</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>71 rows × 6 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-06f9de47-ff8a-4f44-902d-469b1806dda5')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-06f9de47-ff8a-4f44-902d-469b1806dda5 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-06f9de47-ff8a-4f44-902d-469b1806dda5');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-7b2e4bc5-4a40-40ef-88c1-e0b53ad26ff8\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-7b2e4bc5-4a40-40ef-88c1-e0b53ad26ff8')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-7b2e4bc5-4a40-40ef-88c1-e0b53ad26ff8 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "  <div id=\"id_40c3c503-41f9-4579-93bb-06efc26062ba\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('metadata')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_40c3c503-41f9-4579-93bb-06efc26062ba button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('metadata');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "metadata",
              "summary": "{\n  \"name\": \"metadata\",\n  \"rows\": 71,\n  \"fields\": [\n    {\n      \"column\": \"Sample_id\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 71,\n        \"samples\": [\n          \"IN1\",\n          \"F1\",\n          \"ST4\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Novogene ID\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 71,\n        \"samples\": [\n          \"FKDN250100735-1A\",\n          \"FKDN250100713-1A\",\n          \"FKDN250100763-1A\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Sample Name in Report\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 12,\n        \"samples\": [\n          \"HET10\",\n          \"WT22\",\n          \"FHET10\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Genotype\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          \"WT\",\n          \"HET\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Time\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 6,\n        \"min\": 10,\n        \"max\": 22,\n        \"num_unique_values\": 2,\n        \"samples\": [\n          22,\n          10\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Location\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 3,\n        \"samples\": [\n          \"Faecal \",\n          \"Ileum \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Group\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 4,\n        \"samples\": [\n          \"WT_10\",\n          \"HET_22\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 117
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Tree"
      ],
      "metadata": {
        "id": "Ua0AA1DBnDlW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!sudo apt-get install muscle"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "81PpAqaFnFDI",
        "outputId": "b9305096-bcc7-4471-8e8e-eeaeaa577a63"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "muscle is already the newest version (1:3.8.1551-2build1).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 35 not upgraded.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!muscle -in /content/zotus_nochime.fa -out /content/zotus_align.fa -maxiters 1 -diags1"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "u1fa7urYnGHR",
        "outputId": "63160c2c-a1b1-438f-b16a-3114023a0a4c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "MUSCLE v3.8.1551 by Robert C. Edgar\n",
            "\n",
            "http://www.drive5.com/muscle\n",
            "This software is donated to the public domain.\n",
            "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
            "\n",
            "zotus_nochime 1504 seqs, lengths min 400, max 400, avg 400\n",
            "00:00:16     37 MB(9%)  Iter   1  100.00%  K-mer dist pass 1\n",
            "00:00:17     37 MB(9%)  Iter   1  100.00%  K-mer dist pass 2\n",
            "00:00:22   392 MB(97%)  Iter   1  100.00%  Align node\n",
            "00:00:22   395 MB(98%)  Iter   1  100.00%  Root alignment\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!sudo apt-get install fasttree"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "q78H5zQenHfO",
        "outputId": "71018893-df00-4477-99ce-b5eee1b0f022"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "fasttree is already the newest version (2.1.11-2).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 35 not upgraded.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!FastTree /content/zotus_align.fa > tree.nwk"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nmROVTNBnIx9",
        "outputId": "3e5064cc-4bfa-4535-8ec5-5ecb5c83cd5e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "FastTree Version 2.1.11 Double precision (No SSE3)\n",
            "Alignment: /content/zotus_align.fa\n",
            "Amino acid distances: BLOSUM45 Joins: balanced Support: SH-like 1000\n",
            "Search: Normal +NNI +SPR (2 rounds range 10) +ML-NNI opt-each=1\n",
            "TopHits: 1.00*sqrtN close=default refresh=0.80\n",
            "ML Model: Jones-Taylor-Thorton, CAT approximation with 20 rate categories\n",
            "WARNING! 100.0% NUCLEOTIDE CHARACTERS -- IS THIS REALLY A PROTEIN ALIGNMENT?\n",
            "Initial topology in 7.94 seconds\n",
            "Refining topology: 42 rounds ME-NNIs, 2 rounds ME-SPRs, 21 rounds ML-NNIs\n",
            "Total branch-length 83.736 after 32.34 sec\n",
            "ML-NNI round 1: LogLk = -194911.045 NNIs 321 max delta 25.48 Time 108.35\n",
            "Switched to using 20 rate categories (CAT approximation)\n",
            "Rate categories were divided by 0.981 so that average rate = 1.0\n",
            "CAT-based log-likelihoods may not be comparable across runs\n",
            "Use -gamma for approximate but comparable Gamma(20) log-likelihoods\n",
            "ML-NNI round 2: LogLk = -181241.528 NNIs 226 max delta 15.55 Time 166.96\n",
            "ML-NNI round 3: LogLk = -181152.045 NNIs 136 max delta 7.47 Time 192.61\n",
            "ML-NNI round 4: LogLk = -181103.870 NNIs 86 max delta 13.05 Time 204.42\n",
            "ML-NNI round 5: LogLk = -181082.286 NNIs 56 max delta 12.12 Time 212.75\n",
            "ML-NNI round 6: LogLk = -181068.378 NNIs 43 max delta 4.96 Time 216.74\n",
            "ML-NNI round 7: LogLk = -181068.225 NNIs 21 max delta 0.00 Time 219.01\n",
            "Turning off heuristics for final round of ML NNIs (converged)\n",
            "ML-NNI round 8: LogLk = -180947.875 NNIs 161 max delta 13.43 Time 280.50 (final)\n",
            "Optimize all lengths: LogLk = -180944.801 Time 296.53\n",
            "Total time: 349.76 seconds Unique: 1504/1504 Bad splits: 2/1501 Worst delta-LogLk 1.269\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from skbio import TreeNode\n",
        "\n",
        "tree = TreeNode.read('/content/tree.nwk')\n",
        "tree\n",
        "rooted_tree = tree.root_at_midpoint()\n",
        "\n",
        "\"\"\"\n",
        "import pandas as pd\n",
        "metadata = pd.read_csv('Hep_mouse.txt', sep='\\t', index_col='Sample_id')\n",
        "otu_table = pd.read_csv('unoise3_zotu_table.txt', sep='\\t', index_col=0)\n",
        "taxonomy = pd.read_csv('sintax.txt', sep='\\t', index_col=0, header=None)\n",
        "taxonomy.columns = ['raw_sintax', 'hit', 'parsed']\n",
        "\n",
        "rooted_tree = tree.root_at_midpoint()\n",
        "\"\"\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 52
        },
        "id": "j-5QmFBLnKR7",
        "outputId": "4372f4e9-7d82-446e-de0a-a20a28abe2c3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "\"\\nimport pandas as pd\\nmetadata = pd.read_csv('Hep_mouse.txt', sep='\\t', index_col='Sample_id')\\notu_table = pd.read_csv('unoise3_zotu_table.txt', sep='\\t', index_col=0)\\ntaxonomy = pd.read_csv('sintax.txt', sep='\\t', index_col=0, header=None)\\ntaxonomy.columns = ['raw_sintax', 'hit', 'parsed']\\n\\nrooted_tree = tree.root_at_midpoint()\\n\""
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 122
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Alpha + Beta diversity"
      ],
      "metadata": {
        "id": "D8yqA1fvlZB-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from skbio.diversity import alpha_diversity\n",
        "from skbio.diversity import beta_diversity\n",
        "from skbio.stats.ordination import pcoa"
      ],
      "metadata": {
        "id": "hE9IBvbAliXV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from skbio.stats.distance import permanova"
      ],
      "metadata": {
        "id": "oSiQBIkw5yzs"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# faecal\n",
        "faecal_ids   = metadata.loc[metadata['Location'].str.strip() == 'Faecal'].index\n",
        "rare_faecal  = rare_table.filter(faecal_ids, axis='sample', inplace=False)\n",
        "\n",
        "# feature_dist_F  = beta_diversity('jaccard', rare_faecal)\n",
        "feature_dist_wF = beta_diversity('weighted_unifrac', rare_faecal, tree=rooted_tree)\n",
        "# feature_dist_uwF = beta_diversity('unweighted_unifrac', rare_faecal, tree=rooted_tree)\n",
        "# feature_dist_bcF = beta_diversity('braycurtis', rare_faecal)\n",
        "\n",
        "# pc_F            = pcoa(feature_dist_F, number_of_dimensions=10)\n",
        "pc_wF           = pcoa(feature_dist_wF, number_of_dimensions=10)\n",
        "# pc_uwF          = pcoa(feature_dist_uwF, number_of_dimensions=10)\n",
        "# pc_bcF          = pcoa(feature_dist_bcF, number_of_dimensions=10)\n",
        "\n",
        "# metadata_jac_F  = metadata.reindex(pc_F.samples.index)\n",
        "metadata_wF = metadata.reindex(pc_wF.samples.index)\n",
        "# metadata_uwF = metadata.reindex(pc_uwF.samples.index)\n",
        "# metadata_bcF = metadata.reindex(pc_bcF.samples.index)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_l2CloZfBTnX",
        "outputId": "896e8502-015b-4794-a3dd-6a25b134d1c8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/skbio/stats/ordination/_principal_coordinate_analysis.py:214: RuntimeWarning: The result contains negative eigenvalues that are large in magnitude, which may suggest result inaccuracy. See Notes for details. The negative-most eigenvalue is -0.08980241165175605 whereas the largest positive one is 0.8309681738305126.\n",
            "  warn(\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# ileum\n",
        "ileum_ids   = metadata.loc[metadata['Location'].str.strip() == 'Ileum'].index\n",
        "rare_ileum  = rare_table.filter(ileum_ids, axis='sample', inplace=False)\n",
        "\n",
        "# feature_dist_I  = beta_diversity('jaccard', rare_ileum)\n",
        "feature_dist_wI = beta_diversity('weighted_unifrac', rare_ileum, tree=rooted_tree)\n",
        "# feature_dist_uWI = beta_diversity('unweighted_unifrac', rare_ileum, tree=rooted_tree)\n",
        "# feature_dist_bcI = beta_diversity('braycurtis', rare_ileum)\n",
        "\n",
        "# pc_I            = pcoa(feature_dist_I, number_of_dimensions=10)\n",
        "pc_wI= pcoa(feature_dist_wI, number_of_dimensions=10)\n",
        "# pc_uWI          = pcoa(feature_dist_uWI, number_of_dimensions=10)\n",
        "# pc_bcI          = pcoa(feature_dist_bcI, number_of_dimensions=10)\n",
        "\n",
        "# metadata_jac_I  = metadata.reindex(pc_I.samples.index)\n",
        "metadata_wI = metadata.reindex(pc_wI.samples.index)\n",
        "# metadata_uWI = metadata.reindex(pc_uWI.samples.index)\n",
        "# metadata_bcI = metadata.reindex(pc_bcI.samples.index)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "u7KtTurTBV7g",
        "outputId": "effaa4a1-719f-40cf-b3a1-108645cfdfa0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/skbio/stats/ordination/_principal_coordinate_analysis.py:214: RuntimeWarning: The result contains negative eigenvalues that are large in magnitude, which may suggest result inaccuracy. See Notes for details. The negative-most eigenvalue is -0.08073912306732374 whereas the largest positive one is 1.6331770362830527.\n",
            "  warn(\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# duodenum\n",
        "duodenum_ids   = metadata.loc[metadata['Location'].str.strip() == 'Duodenum'].index\n",
        "rare_duodenum  = rare_table.filter(duodenum_ids, axis='sample', inplace=False)\n",
        "\n",
        "# feature_dist_D  = beta_diversity('jaccard', rare_duodenum)\n",
        "feature_dist_wD = beta_diversity('weighted_unifrac', rare_duodenum, tree=rooted_tree)\n",
        "# feature_dist_uWD = beta_diversity('unweighted_unifrac', rare_duodenum, tree=rooted_tree)\n",
        "# feature_dist_bcD = beta_diversity('braycurtis', rare_duodenum)\n",
        "\n",
        "# pc_D            = pcoa(feature_dist_D, number_of_dimensions=10)\n",
        "pc_wD           = pcoa(feature_dist_wD, number_of_dimensions=10)\n",
        "# pc_uWD          = pcoa(feature_dist_uWD, number_of_dimensions=10)\n",
        "# pc_bcD          = pcoa(feature_dist_bcD, number_of_dimensions=10)\n",
        "\n",
        "# metadata_jac_D  = metadata.reindex(pc_D.samples.index)\n",
        "metadata_wD = metadata.reindex(pc_wD.samples.index)\n",
        "# metadata_uWD = metadata.reindex(pc_uWD.samples.index)\n",
        "# metadata_bcD = metadata.reindex(pc_bcD.samples.index)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "YCCZ_fFeBYVt",
        "outputId": "7f737615-29f2-4448-a79c-b4006fd1fdd6"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/skbio/stats/ordination/_principal_coordinate_analysis.py:214: RuntimeWarning: The result contains negative eigenvalues that are large in magnitude, which may suggest result inaccuracy. See Notes for details. The negative-most eigenvalue is -0.05404678046127414 whereas the largest positive one is 1.1801239693030936.\n",
            "  warn(\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Shannon + inv simpson"
      ],
      "metadata": {
        "id": "l95ut0CZlb7n"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "res_alpha = alpha_diversity('shannon', rare_table)\n",
        "alpha_d=pd.DataFrame(res_alpha, columns=['shannon'])\n",
        "res_alpha = pd.merge(alpha_d, metadata, left_index=True, right_index=True)\n",
        "res_alpha.columns # global, for plot later\n",
        "\n",
        "\n",
        "# inv simpson\n",
        "res_alpha_inv = alpha_diversity('inv_simpson', rare_table)\n",
        "alpha_d=pd.DataFrame(res_alpha_inv, columns=['inv_simpson'])\n",
        "res_alpha_inv = pd.merge(alpha_d, metadata, left_index=True, right_index=True)\n",
        "res_alpha_inv.columns"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "o8INfLY7lgEj",
        "outputId": "c4b41046-4336-4c57-ce91-2604f707fae1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Index(['inv_simpson', 'Novogene ID', 'Sample Name in Report', 'Genotype',\n",
              "       'Time', 'Location', 'Group'],\n",
              "      dtype='object')"
            ]
          },
          "metadata": {},
          "execution_count": 128
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## permanova"
      ],
      "metadata": {
        "id": "F1pS2XhZLdV4"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from skbio.stats.distance import permanova\n",
        "import pandas as pd\n",
        "\n",
        "permanova_results = {}\n",
        "\n",
        "# Faecal\n",
        "grouping_F = metadata_wF['Group']\n",
        "permanova_F = permanova(feature_dist_wF, grouping_F, permutations=999)\n",
        "permanova_results['Faecal'] = permanova_F\n",
        "print(f\"Faecal PERMANOVA:\\n{permanova_F}\\n\")\n",
        "\n",
        "# Ileum\n",
        "grouping_I = metadata_wI['Group']\n",
        "permanova_I = permanova(feature_dist_wI, grouping_I, permutations=999)\n",
        "permanova_results['Ileum'] = permanova_I\n",
        "print(f\"Ileum PERMANOVA:\\n{permanova_I}\\n\")\n",
        "\n",
        "# Duodenum\n",
        "grouping_D = metadata_wD['Group']\n",
        "permanova_D = permanova(feature_dist_wD, grouping_D, permutations=999)\n",
        "permanova_results['Duodenum'] = permanova_D\n",
        "print(f\"Duodenum PERMANOVA:\\n{permanova_D}\\n\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "q0oW5VaZH0EP",
        "outputId": "6a842b0a-4108-4f68-a8c2-e179a2c623eb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Faecal PERMANOVA:\n",
            "method name               PERMANOVA\n",
            "test statistic name        pseudo-F\n",
            "sample size                      31\n",
            "number of groups                  4\n",
            "test statistic             1.775878\n",
            "p-value                        0.09\n",
            "number of permutations          999\n",
            "Name: PERMANOVA results, dtype: object\n",
            "\n",
            "Ileum PERMANOVA:\n",
            "method name               PERMANOVA\n",
            "test statistic name        pseudo-F\n",
            "sample size                      24\n",
            "number of groups                  4\n",
            "test statistic             0.783262\n",
            "p-value                       0.656\n",
            "number of permutations          999\n",
            "Name: PERMANOVA results, dtype: object\n",
            "\n",
            "Duodenum PERMANOVA:\n",
            "method name               PERMANOVA\n",
            "test statistic name        pseudo-F\n",
            "sample size                      16\n",
            "number of groups                  4\n",
            "test statistic             1.224708\n",
            "p-value                       0.287\n",
            "number of permutations          999\n",
            "Name: PERMANOVA results, dtype: object\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## pairwise permanova (R)"
      ],
      "metadata": {
        "id": "DTRDaSlVLoDT"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "feature_df_wF=feature_dist_wF.to_data_frame()\n",
        "feature_df_wI=feature_dist_wI.to_data_frame()\n",
        "feature_df_wD=feature_dist_wD.to_data_frame()\n",
        "\n",
        "feature_df_wF.to_csv('feature_dist_wF.txt', sep='\\t')\n",
        "feature_df_wI.to_csv('feature_dist_wI.txt', sep='\\t')\n",
        "feature_df_wD.to_csv('feature_dist_wD.txt', sep='\\t')"
      ],
      "metadata": {
        "id": "KpKJpWTFN9Tq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%load_ext rpy2.ipython"
      ],
      "metadata": {
        "id": "F8b9YDRELqTg",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "1e48428d-76b8-4591-9cf8-34b94d3e96a5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "The rpy2.ipython extension is already loaded. To reload it, use:\n",
            "  %reload_ext rpy2.ipython\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "install.packages(\"vegan\")\n",
        "library(devtools)\n",
        "\n",
        "install_github(\"pmartinezarbizu/pairwiseAdonis/pairwiseAdonis\")\n",
        "\n",
        "library(pairwiseAdonis)\n",
        "library(vegan)\n",
        "library(tidyverse)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 225
        },
        "id": "ZK9_387ILtQ5",
        "outputId": "45cc2136-0e82-4b3e-cd10-256f47100fd6"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Installing package into ‘/usr/local/lib/R/site-library’\n",
              "(as ‘lib’ is unspecified)\n",
              "trying URL 'https://cran.rstudio.com/src/contrib/vegan_2.6-10.tar.gz'\n",
              "Content type 'application/x-gzip' length 1502105 bytes (1.4 MB)\n",
              "==================================================\n",
              "downloaded 1.4 MB\n",
              "\n",
              "\n",
              "The downloaded source packages are in\n",
              "\t‘/tmp/RtmpyIlbXo/downloaded_packages’\n",
              "Skipping install of 'pairwiseAdonis' from a github remote, the SHA1 (cb190f76) has not changed since last install.\n",
              "  Use `force = TRUE` to force installation\n"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "metadata_full<-read_tsv(\"Hep_mouse.txt\")\n",
        "metadata_full$Location <- trimws(metadata_full$Location)\n",
        "metadata_full$Group <- trimws(metadata_full$Group)\n",
        "\n",
        "# read faecal distance matrix\n",
        "w_unifrac_wF <- read.table(\"/content/feature_dist_wF.txt\", sep=\"\\t\", header=TRUE, row.names=1)\n",
        "# get the sample IDs from the matrix\n",
        "faecal_samples <- rownames(w_unifrac_wF)\n",
        "\n",
        "# keep only the faecal samples\n",
        "metadata_wF_R <- metadata_full[match(faecal_samples, metadata_full$Sample_id), ]\n",
        "\n",
        "print(paste(\"Rows in distance matrix:\", nrow(w_unifrac_wF)))\n",
        "print(paste(\"Rows in filtered metadata:\", nrow(metadata_wF_R)))\n",
        "\n",
        "print(head(data.frame(DistMatrixSamples = rownames(w_unifrac_wF), MetadataSamples = metadata_wF_R$Sample_id)))\n",
        "\n",
        "\n",
        "# repeat for ileum\n",
        "w_unifrac_wI<-read.table(\"/content/feature_dist_wI.txt\", sep=\"\\t\", header=TRUE, row.names=1)\n",
        "ileum_samples <- rownames(w_unifrac_wI)\n",
        "\n",
        "metadata_wI_R <- metadata_full[match(ileum_samples, metadata_full$Sample_id), ]\n",
        "\n",
        "print(paste(\"Rows in distance matrix:\", nrow(w_unifrac_wI)))\n",
        "print(paste(\"Rows in filtered metadata:\", nrow(metadata_wI_R)))\n",
        "\n",
        "print(head(data.frame(DistMatrixSamples = rownames(w_unifrac_wI), MetadataSamples = metadata_wI_R$Sample_id)))\n",
        "\n",
        "\n",
        "# repeat for duo\n",
        "w_unifrac_wD<-read.table(\"/content/feature_dist_wD.txt\", sep=\"\\t\", header=TRUE, row.names=1)\n",
        "duo_samples <- rownames(w_unifrac_wD)\n",
        "\n",
        "metadata_wD_R <- metadata_full[match(duo_samples, metadata_full$Sample_id), ]\n",
        "\n",
        "print(paste(\"Rows in distance matrix:\", nrow(w_unifrac_wD)))\n",
        "print(paste(\"Rows in filtered metadata:\", nrow(metadata_wD_R)))\n",
        "\n",
        "print(head(data.frame(DistMatrixSamples = rownames(w_unifrac_wD), MetadataSamples = metadata_wD_R$Sample_id)))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "azeH8iN9LtdR",
        "outputId": "9478eb21-d3ac-459f-f487-99dc8c255e80"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Rows: 71 Columns: 7\n",
            "── Column specification ────────────────────────────────────────────────────────\n",
            "Delimiter: \"\\t\"\n",
            "chr (6): Sample_id, Novogene ID, Sample Name in Report, Genotype, Location, ...\n",
            "dbl (1): Time\n",
            "\n",
            "ℹ Use `spec()` to retrieve the full column specification for this data.\n",
            "ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
            "[1] \"Rows in distance matrix: 31\"\n",
            "[1] \"Rows in filtered metadata: 31\"\n",
            "  DistMatrixSamples MetadataSamples\n",
            "1                F1              F1\n",
            "2               F10             F10\n",
            "3               F11             F11\n",
            "4               F12             F12\n",
            "5               F13             F13\n",
            "6               F14             F14\n",
            "[1] \"Rows in distance matrix: 24\"\n",
            "[1] \"Rows in filtered metadata: 24\"\n",
            "  DistMatrixSamples MetadataSamples\n",
            "1               IN1             IN1\n",
            "2              IN10            IN10\n",
            "3              IN11            IN11\n",
            "4              IN12            IN12\n",
            "5              IN13            IN13\n",
            "6              IN14            IN14\n",
            "[1] \"Rows in distance matrix: 16\"\n",
            "[1] \"Rows in filtered metadata: 16\"\n",
            "  DistMatrixSamples MetadataSamples\n",
            "1               ST1             ST1\n",
            "2              ST10            ST10\n",
            "3              ST11            ST11\n",
            "4              ST12            ST12\n",
            "5              ST13            ST13\n",
            "6              ST14            ST14\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "data.frame(rownames(w_unifrac_wF), metadata_wF_R$Sample_id)\n",
        "data.frame(rownames(w_unifrac_wI), metadata_wI_R$Sample_id)\n",
        "data.frame(rownames(w_unifrac_wD), metadata_wD_R$Sample_id)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "vw0gEYFCLttL",
        "outputId": "dad69769-2053-4451-a687-544b886df67c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "   rownames.w_unifrac_wD. metadata_wD_R.Sample_id\n",
            "1                     ST1                     ST1\n",
            "2                    ST10                    ST10\n",
            "3                    ST11                    ST11\n",
            "4                    ST12                    ST12\n",
            "5                    ST13                    ST13\n",
            "6                    ST14                    ST14\n",
            "7                    ST15                    ST15\n",
            "8                    ST16                    ST16\n",
            "9                     ST2                     ST2\n",
            "10                    ST3                     ST3\n",
            "11                    ST4                     ST4\n",
            "12                    ST5                     ST5\n",
            "13                    ST6                     ST6\n",
            "14                    ST7                     ST7\n",
            "15                    ST8                     ST8\n",
            "16                    ST9                     ST9\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "metadata_wF_R$Group <- factor(metadata_wF_R$Group)\n",
        "\n",
        "results_wF <- pairwise.adonis(x = w_unifrac_wF,factors = metadata_wF_R$Group)\n",
        "\n",
        "print(results_wF)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SLEV_E4MO4-P",
        "outputId": "bf6cad6a-5301-43b1-9e79-a28239c27c9c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "             pairs Df   SumsOfSqs   F.Model         R2 p.value p.adjusted sig\n",
            "1  HET_10 vs WT_10  1 0.038166765 1.7842274 0.12068452   0.124      0.744    \n",
            "2 HET_10 vs HET_22  1 0.006687986 0.2125666 0.01608821   0.934      1.000    \n",
            "3  HET_10 vs WT_22  1 0.022439590 0.7391526 0.06296473   0.525      1.000    \n",
            "4  WT_10 vs HET_22  1 0.055488319 2.5109539 0.13564692   0.045      0.270    \n",
            "5   WT_10 vs WT_22  1 0.114164302 5.7389462 0.29074228   0.007      0.042   .\n",
            "6  HET_22 vs WT_22  1 0.014668564 0.5015698 0.03458728   0.706      1.000    \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "metadata_wI_R$Group <- factor(metadata_wI_R$Group)\n",
        "\n",
        "results_wI <- pairwise.adonis(x = w_unifrac_wI,factors = metadata_wI_R$Group)\n",
        "\n",
        "print(results_wI)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "evggzZQxUGIM",
        "outputId": "0aa3049f-8bc5-4e51-e3d0-e8ce628434d9"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "             pairs Df  SumsOfSqs   F.Model         R2 p.value p.adjusted sig\n",
            "1  WT_22 vs HET_22  1 0.01688326 0.6430035 0.06668083   0.517          1    \n",
            "2  WT_22 vs HET_10  1 0.00997031 0.2997134 0.03222824   0.771          1    \n",
            "3   WT_22 vs WT_10  1 0.01006383 0.3622490 0.02930284   0.777          1    \n",
            "4 HET_22 vs HET_10  1 0.02436574 0.9998410 0.11109541   0.418          1    \n",
            "5  HET_22 vs WT_10  1 0.01650932 0.7932320 0.06726163   0.500          1    \n",
            "6  HET_10 vs WT_10  1 0.01012461 0.3813758 0.03350876   0.846          1    \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "metadata_wD_R$Group <- factor(metadata_wD_R$Group)\n",
        "\n",
        "results_wD <- pairwise.adonis(x = w_unifrac_wD,factors = metadata_wD_R$Group)\n",
        "\n",
        "print(results_wD)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 173
        },
        "id": "FdL5zWluUNHx",
        "outputId": "2272d5e1-7bad-42d1-e6ed-647cf9bb78c9"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "             pairs Df  SumsOfSqs   F.Model         R2 p.value p.adjusted sig\n",
            "1  WT_10 vs HET_10  1 0.01690682 0.4999643 0.06666223   0.699      1.000    \n",
            "2  WT_10 vs HET_22  1 0.05928425 2.1125458 0.23182827   0.096      0.576    \n",
            "3   WT_10 vs WT_22  1 0.03924947 1.5213338 0.20226916   0.280      1.000    \n",
            "4 HET_10 vs HET_22  1 0.01116503 0.3243647 0.05128811   0.943      1.000    \n",
            "5  HET_10 vs WT_22  1 0.02506807 0.7601774 0.13197117   0.472      1.000    \n",
            "6  HET_22 vs WT_22  1 0.03492108 1.4012010 0.21889658   0.245      1.000    \n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Set of permutations < 'minperm'. Generating entire set.\n",
              "Set of permutations < 'minperm'. Generating entire set.\n"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from scipy.stats import mannwhitneyu\n",
        "from statsmodels.stats.multicomp import MultiComparison\n",
        "\n",
        "res_alpha['Location'] = res_alpha['Location'].str.strip()\n",
        "res_alpha['Group'] = res_alpha['Group'].str.strip()\n",
        "\n",
        "location_to_filter = 'Faecal'\n",
        "res_alpha_F = res_alpha[res_alpha['Location'] == location_to_filter].copy()\n",
        "\n",
        "multi_comp_shannon_F = MultiComparison(res_alpha_F['shannon'], res_alpha_F['Group'])\n",
        "result_shannon_F = multi_comp_shannon_F.allpairtest(mannwhitneyu, method='Holm')\n",
        "\n",
        "print(result_shannon_F[0])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "u_kvySFEhiPb",
        "outputId": "b99bc91a-f4ff-44d6-bccb-d96b4e65bd23"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Test Multiple Comparison mannwhitneyu \n",
            "FWER=0.05 method=Holm\n",
            "alphacSidak=0.01, alphacBonf=0.008\n",
            "==========================================\n",
            "group1 group2 stat  pval  pval_corr reject\n",
            "------------------------------------------\n",
            "HET_10 HET_22 23.0 0.6889       1.0  False\n",
            "HET_10  WT_10 29.0 0.8639       1.0  False\n",
            "HET_10  WT_22 13.0 0.2949       1.0  False\n",
            "HET_22  WT_10 58.0 0.1333    0.6666  False\n",
            "HET_22  WT_22 22.0  0.351       1.0  False\n",
            " WT_10  WT_22 12.0 0.0418    0.2507  False\n",
            "------------------------------------------\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "location_to_filter = 'Ileum'\n",
        "res_alpha_I = res_alpha[res_alpha['Location'] == location_to_filter].copy()\n",
        "\n",
        "multi_comp_shannon_I = MultiComparison(res_alpha_I['shannon'], res_alpha_I['Group'])\n",
        "result_shannon_I = multi_comp_shannon_I.allpairtest(mannwhitneyu, method='Holm')\n",
        "\n",
        "print(result_shannon_I[0])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "r7xDYl8DiGsr",
        "outputId": "68bb7a90-0fbe-4146-fde2-f0805d71e2f6"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Test Multiple Comparison mannwhitneyu \n",
            "FWER=0.05 method=Holm\n",
            "alphacSidak=0.01, alphacBonf=0.008\n",
            "==========================================\n",
            "group1 group2 stat  pval  pval_corr reject\n",
            "------------------------------------------\n",
            "HET_10 HET_22 11.0 0.8413       1.0  False\n",
            "HET_10  WT_10 20.0    1.0       1.0  False\n",
            "HET_10  WT_22 17.0 0.7922       1.0  False\n",
            "HET_22  WT_10 26.0 0.4351       1.0  False\n",
            "HET_22  WT_22 21.0  0.329       1.0  False\n",
            " WT_10  WT_22 26.0 0.8518       1.0  False\n",
            "------------------------------------------\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "location_to_filter = 'Duodenum'\n",
        "res_alpha_D = res_alpha[res_alpha['Location'] == location_to_filter].copy()\n",
        "\n",
        "multi_comp_shannon_D = MultiComparison(res_alpha_D['shannon'], res_alpha_D['Group'])\n",
        "result_shannon_D = multi_comp_shannon_D.allpairtest(mannwhitneyu, method='Holm')\n",
        "\n",
        "print(result_shannon_D[0])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "oONJXN6FiUrm",
        "outputId": "fdd9ef15-9edf-4ca2-ad7e-5ddb8c757ba3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Test Multiple Comparison mannwhitneyu \n",
            "FWER=0.05 method=Holm\n",
            "alphacSidak=0.01, alphacBonf=0.008\n",
            "==========================================\n",
            "group1 group2 stat  pval  pval_corr reject\n",
            "------------------------------------------\n",
            "HET_10 HET_22  7.0 0.8857       1.0  False\n",
            "HET_10  WT_10 14.0 0.4127       1.0  False\n",
            "HET_10  WT_22  6.0    1.0       1.0  False\n",
            "HET_22  WT_10 15.0 0.2857       1.0  False\n",
            "HET_22  WT_22  7.0 0.8571       1.0  False\n",
            " WT_10  WT_22  5.0 0.5714       1.0  False\n",
            "------------------------------------------\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "res_alpha_inv['Location'] = res_alpha_inv['Location'].str.strip()\n",
        "res_alpha_inv['Group']    = res_alpha_inv['Group'].str.strip()\n",
        "\n",
        "location_to_filter = 'Faecal'\n",
        "res_alpha_inv_F = res_alpha_inv[res_alpha_inv['Location'] == location_to_filter].copy()\n",
        "\n",
        "multi_comp_inv_F = MultiComparison(res_alpha_inv_F['inv_simpson'], res_alpha_inv_F['Group'])\n",
        "result_inv_F = multi_comp_inv_F.allpairtest(mannwhitneyu, method='Holm')\n",
        "\n",
        "print(result_inv_F[0])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jTdHSdohiiyZ",
        "outputId": "8a702c64-b7fa-4cd3-9826-41518868318d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Test Multiple Comparison mannwhitneyu \n",
            "FWER=0.05 method=Holm\n",
            "alphacSidak=0.01, alphacBonf=0.008\n",
            "==========================================\n",
            "group1 group2 stat  pval  pval_corr reject\n",
            "------------------------------------------\n",
            "HET_10 HET_22 21.0 0.5287       1.0  False\n",
            "HET_10  WT_10 32.0  0.607       1.0  False\n",
            "HET_10  WT_22 12.0 0.2343       1.0  False\n",
            "HET_22  WT_10 54.0  0.251       1.0  False\n",
            "HET_22  WT_22 21.0 0.2991       1.0  False\n",
            " WT_10  WT_22  8.0 0.0115    0.0692  False\n",
            "------------------------------------------\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "location_to_filter = 'Ileum'\n",
        "res_alpha_inv_I = res_alpha_inv[res_alpha_inv['Location'] == location_to_filter].copy()\n",
        "\n",
        "multi_comp_inv_I = MultiComparison(res_alpha_inv_I['inv_simpson'], res_alpha_inv_I['Group'])\n",
        "result_inv_I = multi_comp_inv_I.allpairtest(mannwhitneyu, method='Holm')\n",
        "\n",
        "print(result_inv_I[0])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_XZA8Uo0j9Fw",
        "outputId": "c7a95dac-940f-4c56-a644-a8dfe5f66680"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Test Multiple Comparison mannwhitneyu \n",
            "FWER=0.05 method=Holm\n",
            "alphacSidak=0.01, alphacBonf=0.008\n",
            "==========================================\n",
            "group1 group2 stat  pval  pval_corr reject\n",
            "------------------------------------------\n",
            "HET_10 HET_22 10.0 0.6905       1.0  False\n",
            "HET_10  WT_10 24.0 0.6216       1.0  False\n",
            "HET_10  WT_22 17.0 0.7922       1.0  False\n",
            "HET_22  WT_10 29.0 0.2222       1.0  False\n",
            "HET_22  WT_22 23.0 0.1775       1.0  False\n",
            " WT_10  WT_22 26.0 0.8518       1.0  False\n",
            "------------------------------------------\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "location_to_filter = 'Duodenum'\n",
        "res_alpha_inv_D = res_alpha_inv[res_alpha_inv['Location'] == location_to_filter].copy()\n",
        "\n",
        "multi_comp_inv_D = MultiComparison(res_alpha_inv_D['inv_simpson'], res_alpha_inv_D['Group'])\n",
        "result_inv_D = multi_comp_inv_D.allpairtest(mannwhitneyu, method='Holm')\n",
        "\n",
        "print(result_inv_D[0])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WAPOoybNkGSm",
        "outputId": "e57d6d5e-dc66-4825-8367-8e381e995ceb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Test Multiple Comparison mannwhitneyu \n",
            "FWER=0.05 method=Holm\n",
            "alphacSidak=0.01, alphacBonf=0.008\n",
            "==========================================\n",
            "group1 group2 stat  pval  pval_corr reject\n",
            "------------------------------------------\n",
            "HET_10 HET_22  7.0 0.8857       1.0  False\n",
            "HET_10  WT_10 14.0 0.4127       1.0  False\n",
            "HET_10  WT_22  7.0 0.8571       1.0  False\n",
            "HET_22  WT_10 15.0 0.2857       1.0  False\n",
            "HET_22  WT_22  8.0 0.6286       1.0  False\n",
            " WT_10  WT_22  5.0 0.5714       1.0  False\n",
            "------------------------------------------\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import seaborn as sns, matplotlib.pyplot as plt\n",
        "\n",
        "# shannon frame\n",
        "clean_shannon = res_alpha.copy()\n",
        "clean_shannon['Location'] = clean_shannon['Location'].str.strip()\n",
        "clean_shannon['Group']    = clean_shannon['Group'].str.strip()\n",
        "\n",
        "# inv simpson frame\n",
        "inv_clean = res_alpha_inv.copy()\n",
        "inv_clean['Location'] = inv_clean['Location'].str.strip()\n",
        "inv_clean['Group']    = inv_clean['Group'].str.strip()\n",
        "\n",
        "grp_order = ['WT_10', 'WT_22', 'HET_10', 'HET_22']\n",
        "panel_lbls = ['A', 'B', 'C', 'D', 'E', 'F' ]\n",
        "\n",
        "\n",
        "base = sns.color_palette('pastel', 8)\n",
        "my_palette = dict(zip(grp_order, [base[i] for i in [2, 0, 3, 4]]))\n",
        "\n",
        "# figure layout\n",
        "fig, axes = plt.subplots(\n",
        "    3, 3, figsize=(15, 16),\n",
        "    gridspec_kw={'width_ratios':[1,1,1],\n",
        "                 'height_ratios':[1,1,1]},\n",
        "    sharey=False)\n",
        "\n",
        "ax_a, ax_b, ax_c = axes[0]\n",
        "ax_d, ax_e, ax_f = axes[1]\n",
        "\n",
        "locs        = ['Faecal', 'Ileum', 'Duodenum']\n",
        "panel_lbls  = ['A', '', '', 'B', '', '', 'C', '', '']\n",
        "\n",
        "# shannon boxplots\n",
        "for ax, loc, label in zip([ax_a, ax_b, ax_c], locs, panel_lbls[:3]):\n",
        "    sub = clean_shannon[clean_shannon['Location'] == loc]\n",
        "    sns.boxplot(x='Group', y='shannon',\n",
        "                hue='Group', hue_order=grp_order,\n",
        "                order=grp_order, palette=my_palette,\n",
        "                legend=False, data=sub, ax=ax)\n",
        "\n",
        "    ax.text(-0.12, 1.08, label, transform=ax.transAxes,\n",
        "            fontweight='bold', fontsize=14, ha='right', va='top')\n",
        "\n",
        "    ax.set_title(f'{loc}')\n",
        "    ax.set_xlabel('')\n",
        "    if ax is not axes[0,0]: ax.set_ylabel(''); ax.set_ylabel('')\n",
        "\n",
        "# inv simpson boxplots\n",
        "for ax, loc, label in zip([ax_d, ax_e, ax_f], locs, panel_lbls[3:]):\n",
        "    sub = inv_clean[inv_clean['Location'] == loc]\n",
        "    sns.boxplot(x='Group', y='inv_simpson',\n",
        "                hue='Group', hue_order=grp_order,\n",
        "                order=grp_order, palette=my_palette,\n",
        "                legend=False, data=sub, ax=ax)\n",
        "\n",
        "    ax.text(-0.12, 1.08, label, transform=ax.transAxes,\n",
        "            fontsize=14, fontweight='bold', ha='right', va='top')\n",
        "\n",
        "    ax.set_xlabel('')\n",
        "    if ax is not axes[1,0]: ax.set_ylabel(''); ax.set_ylabel('')\n",
        "\n",
        "\n",
        "#-----------------\n",
        "# PCoA\n",
        "# pcs_bc  = [pc_bcF,  pc_bcI,  pc_bcD]\n",
        "# meta_bc = [metadata_bcF, metadata_bcI, metadata_bcD]\n",
        "pcs_w   = [pc_wF,   pc_wI,   pc_wD]\n",
        "meta_w  = [metadata_wF, metadata_wI, metadata_wD]\n",
        "\n",
        "import numpy as np\n",
        "from matplotlib.patches import Ellipse\n",
        "from scipy.stats import chi2\n",
        "\n",
        "\n",
        "def add_confidence_ellipse(x, y, ax, n_std=2, trim=0.05, **kwargs):\n",
        "    if trim:\n",
        "        r = np.hypot(x - x.mean(), y - y.mean())\n",
        "        keep = r <= np.quantile(r, 1 - trim)\n",
        "        x, y = x[keep], y[keep]\n",
        "    if x.size < 2: # need at least 2 points or it will be pointless.\n",
        "        return\n",
        "    cov = np.cov(x, y)\n",
        "    if np.linalg.det(cov) == 0:\n",
        "        return\n",
        "    vals, vecs = np.linalg.eigh(cov)\n",
        "    order = vals.argsort()[::-1]\n",
        "    vals, vecs = vals[order], vecs[:, order]\n",
        "\n",
        "    q = np.sqrt(chi2.ppf(0.95, 2)) # 95%, times 2\n",
        "    width, height = 2 * q * np.sqrt(vals)\n",
        "    angle = np.degrees(np.arctan2(*vecs[:,0][::-1]))\n",
        "    ell = Ellipse(xy=(x.mean(), y.mean()),\n",
        "                  width=width, height=height,\n",
        "                  angle=angle, **kwargs)\n",
        "    ax.add_patch(ell)\n",
        "\n",
        "for ax, pc, meta, label, loc in zip(axes[2], pcs_w, meta_w,\n",
        "                                    panel_lbls[6:9], locs):\n",
        "    sns.scatterplot(x=pc.samples.PC1, y=pc.samples.PC2,\n",
        "                    hue=meta.Group, style=meta.Location,\n",
        "                    palette=my_palette, legend=False, ax=ax)\n",
        "\n",
        "    for g in meta.Group.unique():\n",
        "        m = meta.Group == g\n",
        "        add_confidence_ellipse(pc.samples.PC1[m].values,\n",
        "                               pc.samples.PC2[m].values,\n",
        "                               ax=ax, edgecolor=my_palette[g],\n",
        "                               facecolor=my_palette[g], linewidth=1.3, alpha=0.15)\n",
        "\n",
        "    ax.text(-0.12, 1.08, label, transform=ax.transAxes,\n",
        "            fontweight='bold', fontsize=14, ha='right', va='top')\n",
        "    ax.set_xlabel('PCoA1 ({:.1f}%)'.format(pc.proportion_explained[0]*100))\n",
        "    ax.set_ylabel('PCoA2 ({:.1f}%)'.format(pc.proportion_explained[1]*100))\n",
        "\n",
        "    # p value\n",
        "    p_result = permanova_results[loc]\n",
        "    p_val = p_result['p-value']\n",
        "\n",
        "    if p_val < 0.001:\n",
        "      p_text = \"p < 0.001\"\n",
        "    else:\n",
        "      p_text = f\"p = {p_val:.3f}\"\n",
        "    ax.text(0.97, 0.1, f\"PERMANOVA\\n{p_text}\", transform=ax.transAxes, fontsize=9,\n",
        "        verticalalignment='top', horizontalalignment='right',\n",
        "        bbox=dict(boxstyle='round,pad=0.3', fc='white', alpha=0.7))\n",
        "# do not touch the labels\n",
        "\n",
        "plt.tight_layout()\n",
        "plt.show()\n",
        "plt.savefig('fig6_diversity.png')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "YojQMfx8OXgc",
        "outputId": "1a391fcd-33fd-4555-e93c-803a2be14eb1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-179-16f9fbfc1a90>:112: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
            "  ax.set_xlabel('PCoA1 ({:.1f}%)'.format(pc.proportion_explained[0]*100))\n",
            "<ipython-input-179-16f9fbfc1a90>:112: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
            "  ax.set_xlabel('PCoA1 ({:.1f}%)'.format(pc.proportion_explained[0]*100))\n",
            "<ipython-input-179-16f9fbfc1a90>:112: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
            "  ax.set_xlabel('PCoA1 ({:.1f}%)'.format(pc.proportion_explained[0]*100))\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# taxa plots"
      ],
      "metadata": {
        "id": "Re7xcDKpteNu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "# Read taxonomy file\n",
        "sintax = pd.read_csv(\"sintax.txt\", sep='\\t', header=None, usecols=[0, 3], names=['ID', 'TaxonomyDetails'])\n",
        "taxa = sintax.set_index('ID')['TaxonomyDetails'].str.split(',', expand=True)\n",
        "taxa.columns = ['kingdom', 'phylum', 'class', 'order', 'family', 'genus']\n",
        "\n",
        "taxa.to_csv('taxonomy.txt', sep='\\t')\n",
        "feature_table = rare_table.to_dataframe()\n",
        "feature_table.to_csv('rare_table.txt', sep='\\t')\n",
        "\n",
        "# Load saved data\n",
        "metadata = metadata.astype(str)\n",
        "taxa = pd.read_csv('taxonomy.txt', sep='\\t', index_col=0)\n",
        "feature_table = pd.read_csv('rare_table.txt', sep='\\t', index_col=0)"
      ],
      "metadata": {
        "id": "y_fIQ-_EtnTI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## phylum plots"
      ],
      "metadata": {
        "id": "k8FVKtRhtjKf"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Compute relative abundance and group by phylum\n",
        "rel_table = feature_table.div(feature_table.sum()) * 100\n",
        "taxa_results = taxa.merge(rel_table, left_index=True, right_index=True)\n",
        "sumtax = taxa_results.groupby(\"phylum\").sum().T\n",
        "\n",
        "# Ensure metadata and sumtax have the same index type\n",
        "metadata.index = metadata.index.astype(str)\n",
        "sumtax.index = sumtax.index.astype(str)\n",
        "\n",
        "# Merge metadata with sumtax\n",
        "sumtax = metadata.merge(sumtax, left_index=True, right_index=True)"
      ],
      "metadata": {
        "id": "85hjunXktf3Q"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "metadata_column = \"Group\"  # Change this to the column you want to keep\n",
        "phylum_columns = [col for col in sumtax.columns if col.startswith(\"p:\")]\n",
        "sumtax = sumtax[[metadata_column] + phylum_columns]\n",
        "sumtax\n",
        "\n",
        "# metadata_column = \"Location\"  # Change this to the column you want to keep\n",
        "# phylum_columns = [col for col in sumtax.columns if col.startswith(\"p:\")]\n",
        "# sumtax = sumtax[[metadata_column] + phylum_columns]\n",
        "# sumtax"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 443
        },
        "id": "dkFru5SSt20u",
        "outputId": "481e4e95-443a-41db-90c0-3aa9462fdbde"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       Group p:Acidobacteria p:Actinobacteria p:Armatimonadetes  \\\n",
              "F1    HET_10             0.0         2.742862               0.0   \n",
              "F2     WT_10             0.0          6.00838               0.0   \n",
              "F3     WT_22             0.0         1.727787               0.0   \n",
              "F4    HET_10             0.0         0.686795               0.0   \n",
              "F5     WT_10             0.0         3.014989               0.0   \n",
              "...      ...             ...              ...               ...   \n",
              "F28   HET_22             0.0           0.4881               0.0   \n",
              "F29   HET_22             0.0         0.600406               0.0   \n",
              "F30    WT_10             0.0         2.496652               0.0   \n",
              "F31    WT_10             0.0         5.537558               0.0   \n",
              "ST16   WT_10             0.0         1.740746               0.0   \n",
              "\n",
              "     p:Bacteroidetes p:Campilobacterota p:Candidatus_Saccharibacteria  \\\n",
              "F1         43.203317           0.030236                           0.0   \n",
              "F2         28.810851           0.112306                      0.177098   \n",
              "F3          30.52136           0.073431                      0.207334   \n",
              "F4         28.698544           0.587448                       0.07775   \n",
              "F5         37.799663           0.051834                           0.0   \n",
              "...              ...                ...                           ...   \n",
              "F28        42.114811                0.0                      0.004319   \n",
              "F29        59.362447            0.07775                      0.038875   \n",
              "F30        46.455877           0.051834                      0.043195   \n",
              "F31        46.179431           0.047514                      0.038875   \n",
              "ST16        2.894043           0.155501                           0.0   \n",
              "\n",
              "     p:Chlamydiae p:Chloroflexi p:Cyanobacteria/Chloroplast  ...  \\\n",
              "F1            0.0           0.0                         0.0  ...   \n",
              "F2            0.0           0.0                         0.0  ...   \n",
              "F3            0.0           0.0                         0.0  ...   \n",
              "F4            0.0           0.0                         0.0  ...   \n",
              "F5            0.0           0.0                         0.0  ...   \n",
              "...           ...           ...                         ...  ...   \n",
              "F28           0.0           0.0                         0.0  ...   \n",
              "F29           0.0           0.0                         0.0  ...   \n",
              "F30           0.0           0.0                         0.0  ...   \n",
              "F31           0.0           0.0                         0.0  ...   \n",
              "ST16          0.0           0.0                    0.863894  ...   \n",
              "\n",
              "     p:Latescibacteria p:Nitrospirae p:Parcubacteria p:Planctomycetes  \\\n",
              "F1                 0.0           0.0             0.0              0.0   \n",
              "F2                 0.0           0.0             0.0              0.0   \n",
              "F3                 0.0           0.0             0.0              0.0   \n",
              "F4                 0.0           0.0             0.0              0.0   \n",
              "F5                 0.0           0.0             0.0              0.0   \n",
              "...                ...           ...             ...              ...   \n",
              "F28                0.0           0.0             0.0              0.0   \n",
              "F29                0.0           0.0             0.0              0.0   \n",
              "F30                0.0           0.0             0.0              0.0   \n",
              "F31                0.0           0.0             0.0              0.0   \n",
              "ST16               0.0           0.0             0.0              0.0   \n",
              "\n",
              "     p:Plantae p:Proteobacteria p:Rhodothermaeota p:SR1 p:Tenericutes  \\\n",
              "F1         0.0         0.760226               0.0   0.0      0.025917   \n",
              "F2         0.0         0.622003               0.0   0.0      0.064792   \n",
              "F3    0.004319         0.794782               0.0   0.0      0.410349   \n",
              "F4         0.0         1.559328               0.0   0.0      0.613364   \n",
              "F5         0.0         1.209451               0.0   0.0      0.025917   \n",
              "...        ...              ...               ...   ...           ...   \n",
              "F28        0.0         0.120945               0.0   0.0       0.64792   \n",
              "F29        0.0         0.168459               0.0   0.0      1.291521   \n",
              "F30        0.0         0.341238               0.0   0.0      0.228932   \n",
              "F31        0.0         0.393072               0.0   0.0      0.038875   \n",
              "ST16       0.0         2.699667               0.0   0.0     57.340936   \n",
              "\n",
              "     p:Verrucomicrobia  \n",
              "F1                 0.0  \n",
              "F2                 0.0  \n",
              "F3                 0.0  \n",
              "F4                 0.0  \n",
              "F5                 0.0  \n",
              "...                ...  \n",
              "F28                0.0  \n",
              "F29                0.0  \n",
              "F30                0.0  \n",
              "F31                0.0  \n",
              "ST16          0.656559  \n",
              "\n",
              "[71 rows x 25 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-fc8f98fc-776d-49d0-8efe-253c91f10f6c\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Group</th>\n",
              "      <th>p:Acidobacteria</th>\n",
              "      <th>p:Actinobacteria</th>\n",
              "      <th>p:Armatimonadetes</th>\n",
              "      <th>p:Bacteroidetes</th>\n",
              "      <th>p:Campilobacterota</th>\n",
              "      <th>p:Candidatus_Saccharibacteria</th>\n",
              "      <th>p:Chlamydiae</th>\n",
              "      <th>p:Chloroflexi</th>\n",
              "      <th>p:Cyanobacteria/Chloroplast</th>\n",
              "      <th>...</th>\n",
              "      <th>p:Latescibacteria</th>\n",
              "      <th>p:Nitrospirae</th>\n",
              "      <th>p:Parcubacteria</th>\n",
              "      <th>p:Planctomycetes</th>\n",
              "      <th>p:Plantae</th>\n",
              "      <th>p:Proteobacteria</th>\n",
              "      <th>p:Rhodothermaeota</th>\n",
              "      <th>p:SR1</th>\n",
              "      <th>p:Tenericutes</th>\n",
              "      <th>p:Verrucomicrobia</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>F1</th>\n",
              "      <td>HET_10</td>\n",
              "      <td>0.0</td>\n",
              "      <td>2.742862</td>\n",
              "      <td>0.0</td>\n",
              "      <td>43.203317</td>\n",
              "      <td>0.030236</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.760226</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.025917</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F2</th>\n",
              "      <td>WT_10</td>\n",
              "      <td>0.0</td>\n",
              "      <td>6.00838</td>\n",
              "      <td>0.0</td>\n",
              "      <td>28.810851</td>\n",
              "      <td>0.112306</td>\n",
              "      <td>0.177098</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.622003</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.064792</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F3</th>\n",
              "      <td>WT_22</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.727787</td>\n",
              "      <td>0.0</td>\n",
              "      <td>30.52136</td>\n",
              "      <td>0.073431</td>\n",
              "      <td>0.207334</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.004319</td>\n",
              "      <td>0.794782</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.410349</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F4</th>\n",
              "      <td>HET_10</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.686795</td>\n",
              "      <td>0.0</td>\n",
              "      <td>28.698544</td>\n",
              "      <td>0.587448</td>\n",
              "      <td>0.07775</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.559328</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.613364</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F5</th>\n",
              "      <td>WT_10</td>\n",
              "      <td>0.0</td>\n",
              "      <td>3.014989</td>\n",
              "      <td>0.0</td>\n",
              "      <td>37.799663</td>\n",
              "      <td>0.051834</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.209451</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.025917</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F28</th>\n",
              "      <td>HET_22</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.4881</td>\n",
              "      <td>0.0</td>\n",
              "      <td>42.114811</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.004319</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.120945</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.64792</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F29</th>\n",
              "      <td>HET_22</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.600406</td>\n",
              "      <td>0.0</td>\n",
              "      <td>59.362447</td>\n",
              "      <td>0.07775</td>\n",
              "      <td>0.038875</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.168459</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.291521</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F30</th>\n",
              "      <td>WT_10</td>\n",
              "      <td>0.0</td>\n",
              "      <td>2.496652</td>\n",
              "      <td>0.0</td>\n",
              "      <td>46.455877</td>\n",
              "      <td>0.051834</td>\n",
              "      <td>0.043195</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.341238</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.228932</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F31</th>\n",
              "      <td>WT_10</td>\n",
              "      <td>0.0</td>\n",
              "      <td>5.537558</td>\n",
              "      <td>0.0</td>\n",
              "      <td>46.179431</td>\n",
              "      <td>0.047514</td>\n",
              "      <td>0.038875</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.393072</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.038875</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ST16</th>\n",
              "      <td>WT_10</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.740746</td>\n",
              "      <td>0.0</td>\n",
              "      <td>2.894043</td>\n",
              "      <td>0.155501</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.863894</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>2.699667</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>57.340936</td>\n",
              "      <td>0.656559</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>71 rows × 25 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-fc8f98fc-776d-49d0-8efe-253c91f10f6c')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-fc8f98fc-776d-49d0-8efe-253c91f10f6c button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-fc8f98fc-776d-49d0-8efe-253c91f10f6c');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-6189bde8-9950-495c-a04b-fe4bc184c7af\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-6189bde8-9950-495c-a04b-fe4bc184c7af')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-6189bde8-9950-495c-a04b-fe4bc184c7af button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "  <div id=\"id_9a0ff224-6644-4b41-a059-7aa3c3e11e27\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('sumtax')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_9a0ff224-6644-4b41-a059-7aa3c3e11e27 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('sumtax');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "sumtax"
            }
          },
          "metadata": {},
          "execution_count": 147
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "sumtax_F = sumtax[sumtax.index.str.startswith(\"F\")]\n",
        "sumtax_I = sumtax[sumtax.index.str.startswith(\"I\")]\n",
        "sumtax_ST = sumtax[sumtax.index.str.startswith(\"ST\")]"
      ],
      "metadata": {
        "id": "tz4JoAVduVaT"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#Faecal\n",
        "# Aggregate and reshape for visualization\n",
        "taxabar = pd.melt(sumtax_F.groupby('Group').mean().reset_index(), id_vars=['Group'], var_name='Taxa', value_name='Abs')\n",
        "taxabar.loc[taxabar['Abs'] < 1, 'Taxa'] = 'Other'\n",
        "taxabar_aggregated = taxabar.groupby(['Group', 'Taxa']).sum().reset_index()\n",
        "\n",
        "taxabar_pivoted = taxabar_aggregated.pivot(index='Group', columns='Taxa', values='Abs').fillna(0)\n",
        "\n",
        "faecal_phyla_data = taxabar_pivoted.copy() # important!!\n",
        "\n",
        "# Plot stacked bar chart\n",
        "plt.figure(figsize=(10, 6))\n",
        "taxabar_pivoted.plot(kind='bar', stacked=True, color=sns.color_palette(\"Set3\", len(taxabar_pivoted.columns)))\n",
        "plt.xlabel('Group')\n",
        "plt.ylabel('%Abundance')\n",
        "plt.legend(title='Taxa', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
        "plt.tight_layout()\n",
        "\n",
        "plt.savefig('fig4a_faecal.png')\n",
        "\n",
        "\n",
        "\n",
        "#Ileum\n",
        "# Aggregate and reshape for visualization\n",
        "taxabar = pd.melt(sumtax_I.groupby('Group').mean().reset_index(), id_vars=['Group'], var_name='Taxa', value_name='Abs')\n",
        "taxabar.loc[taxabar['Abs'] < 1, 'Taxa'] = 'Other'\n",
        "taxabar_aggregated = taxabar.groupby(['Group', 'Taxa']).sum().reset_index()\n",
        "\n",
        "taxabar_pivoted = taxabar_aggregated.pivot(index='Group', columns='Taxa', values='Abs').fillna(0)\n",
        "\n",
        "ileum_phyla_data = taxabar_pivoted.copy() # important!!\n",
        "\n",
        "# Plot stacked bar chart\n",
        "plt.figure(figsize=(10, 6))\n",
        "taxabar_pivoted.plot(kind='bar', stacked=True, color=sns.color_palette(\"Set3\", len(taxabar_pivoted.columns)))\n",
        "plt.xlabel('Group')\n",
        "plt.ylabel('%Abundance')\n",
        "plt.legend(title='Taxa', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
        "plt.tight_layout()\n",
        "\n",
        "plt.savefig('fig4b_ileum.png')\n",
        "\n",
        "\n",
        "\n",
        "#Duo\n",
        "# Aggregate and reshape for visualization\n",
        "taxabar = pd.melt(sumtax_ST.groupby('Group').mean().reset_index(), id_vars=['Group'], var_name='Taxa', value_name='Abs')\n",
        "taxabar.loc[taxabar['Abs'] < 1, 'Taxa'] = 'Other'\n",
        "taxabar_aggregated = taxabar.groupby(['Group', 'Taxa']).sum().reset_index()\n",
        "\n",
        "taxabar_pivoted = taxabar_aggregated.pivot(index='Group', columns='Taxa', values='Abs').fillna(0)\n",
        "\n",
        "duodenum_phyla_data = taxabar_pivoted.copy() # important!!\n",
        "\n",
        "# Plot stacked bar chart\n",
        "plt.figure(figsize=(10, 6))\n",
        "taxabar_pivoted.plot(kind='bar', stacked=True, color=sns.color_palette(\"Set3\", len(taxabar_pivoted.columns)))\n",
        "plt.xlabel('Group')\n",
        "plt.ylabel('%Abundance')\n",
        "plt.legend(title='Taxa', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
        "plt.tight_layout()\n",
        "\n",
        "plt.savefig('fig4c_duodenum.png')\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_O6gMjPBuaYG",
        "outputId": "3b144c9a-c628-4c5e-a789-17279e835f53"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-149-4edbafe6caae>:7: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
            "  taxabar_pivoted = taxabar_aggregated.pivot(index='Group', columns='Taxa', values='Abs').fillna(0)\n",
            "<ipython-input-149-4edbafe6caae>:29: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
            "  taxabar_pivoted = taxabar_aggregated.pivot(index='Group', columns='Taxa', values='Abs').fillna(0)\n",
            "<ipython-input-149-4edbafe6caae>:51: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
            "  taxabar_pivoted = taxabar_aggregated.pivot(index='Group', columns='Taxa', values='Abs').fillna(0)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## genera plots"
      ],
      "metadata": {
        "id": "LkBvIdcAuqDe"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Compute relative abundance and group by phylum\n",
        "sumtax = taxa_results.groupby(\"genus\").sum().T\n",
        "\n",
        "# Ensure metadata and sumtax have the same index type\n",
        "metadata.index = metadata.index.astype(str)\n",
        "sumtax.index = sumtax.index.astype(str)\n",
        "\n",
        "# Merge metadata with sumtax\n",
        "sumtax = metadata.merge(sumtax, left_index=True, right_index=True)"
      ],
      "metadata": {
        "id": "y_7jp9Biucrj"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "metadata_column = \"Group\"  # Change this to the column you want to keep\n",
        "genus_columns = [col for col in sumtax.columns if col.startswith(\"g:\")]\n",
        "\n",
        "sumtax = sumtax[[metadata_column] + genus_columns]"
      ],
      "metadata": {
        "id": "Z5f1fANFuffb"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "sumtax_F = sumtax[sumtax.index.str.startswith(\"F\")]\n",
        "sumtax_I = sumtax[sumtax.index.str.startswith(\"I\")]\n",
        "sumtax_ST = sumtax[sumtax.index.str.startswith(\"ST\")]"
      ],
      "metadata": {
        "id": "duFkm-TgvAKR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#Faecal\n",
        "# Aggregate and reshape for visualization\n",
        "taxabar = pd.melt(sumtax_F.groupby('Group').mean().reset_index(), id_vars=['Group'], var_name='Taxa', value_name='Abs')\n",
        "taxabar.loc[taxabar['Abs'] < 1, 'Taxa'] = 'Other'\n",
        "taxabar_aggregated = taxabar.groupby(['Group', 'Taxa']).sum().reset_index()\n",
        "\n",
        "taxabar_pivoted = taxabar_aggregated.pivot(index='Group', columns='Taxa', values='Abs').fillna(0)\n",
        "\n",
        "faecal_genera_data = taxabar_pivoted.copy() # important!!\n",
        "\n",
        "# Plot stacked bar chart\n",
        "plt.figure(figsize=(10, 6))\n",
        "taxabar_pivoted.plot(kind='bar', stacked=True, color=sns.color_palette(\"Set3\", len(taxabar_pivoted.columns)))\n",
        "plt.xlabel('Group')\n",
        "plt.ylabel('%Abundance')\n",
        "plt.legend(title='Taxa', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
        "plt.tight_layout()\n",
        "\n",
        "plt.savefig('fig5a_faecal.png')\n",
        "\n",
        "\n",
        "#Ileum\n",
        "# Aggregate and reshape for visualization\n",
        "taxabar = pd.melt(sumtax_I.groupby('Group').mean().reset_index(), id_vars=['Group'], var_name='Taxa', value_name='Abs')\n",
        "taxabar.loc[taxabar['Abs'] < 1, 'Taxa'] = 'Other'\n",
        "taxabar_aggregated = taxabar.groupby(['Group', 'Taxa']).sum().reset_index()\n",
        "\n",
        "taxabar_pivoted = taxabar_aggregated.pivot(index='Group', columns='Taxa', values='Abs').fillna(0)\n",
        "\n",
        "ileum_genera_data = taxabar_pivoted.copy() # important!!\n",
        "\n",
        "# Plot stacked bar chart\n",
        "plt.figure(figsize=(10, 6))\n",
        "taxabar_pivoted.plot(kind='bar', stacked=True, color=sns.color_palette(\"Set3\", len(taxabar_pivoted.columns)))\n",
        "plt.xlabel('Group')\n",
        "plt.ylabel('%Abundance')\n",
        "plt.legend(title='Taxa', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
        "plt.tight_layout()\n",
        "\n",
        "plt.savefig('fig5b_ileum.png')\n",
        "\n",
        "#Duo\n",
        "# Aggregate and reshape for visualization\n",
        "taxabar = pd.melt(sumtax_ST.groupby('Group').mean().reset_index(), id_vars=['Group'], var_name='Taxa', value_name='Abs')\n",
        "taxabar.loc[taxabar['Abs'] < 1, 'Taxa'] = 'Other'\n",
        "taxabar_aggregated = taxabar.groupby(['Group', 'Taxa']).sum().reset_index()\n",
        "\n",
        "taxabar_pivoted = taxabar_aggregated.pivot(index='Group', columns='Taxa', values='Abs').fillna(0)\n",
        "\n",
        "duodenum_genera_data = taxabar_pivoted.copy() # important!!\n",
        "\n",
        "# Plot stacked bar chart\n",
        "plt.figure(figsize=(10, 6))\n",
        "taxabar_pivoted.plot(kind='bar', stacked=True, color=sns.color_palette(\"Set3\", len(taxabar_pivoted.columns)))\n",
        "plt.xlabel('Location')\n",
        "plt.ylabel('%Abundance')\n",
        "plt.legend(title='Taxa', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
        "plt.tight_layout()\n",
        "\n",
        "plt.savefig('fig5c_duodenum.png')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6hUFO-9RvDET",
        "outputId": "51833265-db26-4082-c81d-b0e9191053f6"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-153-3d8a14a7ab89>:7: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
            "  taxabar_pivoted = taxabar_aggregated.pivot(index='Group', columns='Taxa', values='Abs').fillna(0)\n",
            "<ipython-input-153-3d8a14a7ab89>:28: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
            "  taxabar_pivoted = taxabar_aggregated.pivot(index='Group', columns='Taxa', values='Abs').fillna(0)\n",
            "<ipython-input-153-3d8a14a7ab89>:48: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
            "  taxabar_pivoted = taxabar_aggregated.pivot(index='Group', columns='Taxa', values='Abs').fillna(0)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## per-phyla plots"
      ],
      "metadata": {
        "id": "ANPybkG4vLOH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "taxa_grouped = pd.merge(taxa, feature_table, left_index=True, right_index=True)\n",
        "result = q2p.tax_glom_table(taxa_grouped, \"phylum\", sample_indices=range(6, 77)) #Change here for sample number\n",
        "melted_df = pd.melt(result, id_vars=['phylum'], var_name='sample_id', value_name='Value')\n",
        "metadata['sample_id'] = metadata.index\n",
        "merged_df = pd.merge(melted_df, metadata, on='sample_id')\n",
        "\n",
        "merged_df.pivot\n",
        "\n",
        "merged_df_F = merged_df[merged_df[\"sample_id\"].astype(str).str.startswith(\"F\")]\n",
        "merged_df_I = merged_df[merged_df[\"sample_id\"].astype(str).str.startswith(\"I\")]\n",
        "merged_df_ST = merged_df[merged_df[\"sample_id\"].astype(str).str.startswith(\"ST\")]"
      ],
      "metadata": {
        "id": "SE-2-X51vVsc"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "\n",
        "# Faecal\n",
        "# Create a FacetGrid for boxplots\n",
        "g = sns.FacetGrid(merged_df_F, col=\"phylum\", col_wrap=3, height=4, sharey=False)\n",
        "\n",
        "# Map boxplot to each facet\n",
        "g.map_dataframe(sns.boxplot, x=\"Group\", y=\"Value\", showfliers=False, order=sorted(merged_df_F[\"Group\"].unique()))\n",
        "g.map_dataframe(sns.swarmplot, x=\"Group\", y=\"Value\", color=\".25\", order=sorted(merged_df_F[\"Group\"].unique()))\n",
        "\n",
        "g.set_axis_labels(\"Group\", \"Counts\")\n",
        "g.set_titles(\"{col_name}\")\n",
        "\n",
        "g.tight_layout()\n",
        "\n",
        "plt.show()\n",
        "\n",
        "plt.savefig('fig6a_faecal.png')\n",
        "\n",
        "\"\"\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 70
        },
        "id": "dmeVX_v7vf1G",
        "outputId": "b5e985a9-325f-4b89-fbd5-2913bfd724ce"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'\\n\\n# Faecal\\n# Create a FacetGrid for boxplots\\ng = sns.FacetGrid(merged_df_F, col=\"phylum\", col_wrap=3, height=4, sharey=False)\\n\\n# Map boxplot to each facet\\ng.map_dataframe(sns.boxplot, x=\"Group\", y=\"Value\", showfliers=False, order=sorted(merged_df_F[\"Group\"].unique()))\\ng.map_dataframe(sns.swarmplot, x=\"Group\", y=\"Value\", color=\".25\", order=sorted(merged_df_F[\"Group\"].unique()))\\n\\ng.set_axis_labels(\"Group\", \"Counts\")\\ng.set_titles(\"{col_name}\")\\n\\ng.tight_layout()\\n\\nplt.show()\\n\\nplt.savefig(\\'fig6a_faecal.png\\')\\n\\n'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 155
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "\n",
        "# Ileum\n",
        "# Create a FacetGrid for boxplots\n",
        "g = sns.FacetGrid(merged_df_I, col=\"phylum\", col_wrap=3, height=4, sharey=False)\n",
        "\n",
        "# Map boxplot to each facet\n",
        "g.map_dataframe(sns.boxplot, x=\"Group\", y=\"Value\", showfliers=False, order=sorted(merged_df_I[\"Group\"].unique()))\n",
        "g.map_dataframe(sns.swarmplot, x=\"Group\", y=\"Value\", color=\".25\", order=sorted(merged_df_I[\"Group\"].unique()))\n",
        "\n",
        "g.set_axis_labels(\"Group\", \"Counts\")\n",
        "g.set_titles(\"{col_name}\")\n",
        "\n",
        "g.tight_layout()\n",
        "\n",
        "plt.show()\n",
        "\n",
        "plt.savefig('fig6b_ileum.png')\n",
        "\n",
        "\"\"\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 70
        },
        "id": "tS5D7iBjvqEp",
        "outputId": "e3c25256-7967-4b1b-a070-c80f58a30a65"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'\\n\\n# Ileum\\n# Create a FacetGrid for boxplots\\ng = sns.FacetGrid(merged_df_I, col=\"phylum\", col_wrap=3, height=4, sharey=False)\\n\\n# Map boxplot to each facet\\ng.map_dataframe(sns.boxplot, x=\"Group\", y=\"Value\", showfliers=False, order=sorted(merged_df_I[\"Group\"].unique()))\\ng.map_dataframe(sns.swarmplot, x=\"Group\", y=\"Value\", color=\".25\", order=sorted(merged_df_I[\"Group\"].unique()))\\n\\ng.set_axis_labels(\"Group\", \"Counts\")\\ng.set_titles(\"{col_name}\")\\n\\ng.tight_layout()\\n\\nplt.show()\\n\\nplt.savefig(\\'fig6b_ileum.png\\')\\n\\n'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 156
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "\n",
        "# Duo\n",
        "# Create a FacetGrid for boxplots\n",
        "g = sns.FacetGrid(merged_df_ST, col=\"phylum\", col_wrap=3, height=4, sharey=False)\n",
        "\n",
        "# Map boxplot to each facet\n",
        "g.map_dataframe(sns.boxplot, x=\"Group\", y=\"Value\", showfliers=False, order=sorted(merged_df_ST[\"Group\"].unique()))\n",
        "g.map_dataframe(sns.swarmplot, x=\"Group\", y=\"Value\", color=\".25\", order=sorted(merged_df_ST[\"Group\"].unique()))\n",
        "\n",
        "g.set_axis_labels(\"Group\", \"Counts\")\n",
        "g.set_titles(\"{col_name}\")\n",
        "\n",
        "g.tight_layout()\n",
        "\n",
        "plt.show()\n",
        "\n",
        "plt.savefig('fig6c_duodenum.png')\n",
        "\n",
        "\"\"\""
      ],
      "metadata": {
        "id": "FuaYUR42vsls",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 70
        },
        "outputId": "55fd3e1e-357b-4203-cf89-2d8d0bf41999"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'\\n\\n# Duo\\n# Create a FacetGrid for boxplots\\ng = sns.FacetGrid(merged_df_ST, col=\"phylum\", col_wrap=3, height=4, sharey=False)\\n\\n# Map boxplot to each facet\\ng.map_dataframe(sns.boxplot, x=\"Group\", y=\"Value\", showfliers=False, order=sorted(merged_df_ST[\"Group\"].unique()))\\ng.map_dataframe(sns.swarmplot, x=\"Group\", y=\"Value\", color=\".25\", order=sorted(merged_df_ST[\"Group\"].unique()))\\n\\ng.set_axis_labels(\"Group\", \"Counts\")\\ng.set_titles(\"{col_name}\")\\n\\ng.tight_layout()\\n\\nplt.show()\\n\\nplt.savefig(\\'fig6c_duodenum.png\\')\\n\\n'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 157
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### heatmap (phylum)"
      ],
      "metadata": {
        "id": "vrNjOrgnv_hW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# heatmap of log2 fold change (ZT22 vs ZT10) in phylum... consider separating?\n",
        "# normalised for each location and change type (WT vs HET). Z-score\n",
        "\n",
        "from scipy.stats import zscore\n",
        "import matplotlib.pyplot as plt\n",
        "import matplotlib.gridspec as gridspec\n",
        "import matplotlib.cm as cm\n",
        "\n",
        "\n",
        "# fetch copies instead. avoid crashing\n",
        "locations_data = {\n",
        "    'Faecal': merged_df_F.copy() if 'merged_df_F' in locals() and merged_df_F is not None else None,\n",
        "    'Ileum': merged_df_I.copy() if 'merged_df_I' in locals() and merged_df_I is not None else None,\n",
        "    'Duodenum': merged_df_ST.copy() if 'merged_df_ST' in locals() and merged_df_ST is not None else None\n",
        "}\n",
        "\n",
        "# load and calculate\n",
        "processed_data = {}\n",
        "all_phyla = set()\n",
        "\n",
        "# group names\n",
        "wt_groups = ['WT_10', 'WT_22']\n",
        "het_groups = ['HET_10', 'HET_22']\n",
        "required_groups = wt_groups + het_groups\n",
        "\n",
        "for loc_name, loc_df in locations_data.items():\n",
        "    if loc_df is None or loc_df.empty:\n",
        "        print(f\"Skipping {loc_name}: No data.\")\n",
        "        continue\n",
        "\n",
        "    loc_df['Value'] = pd.to_numeric(loc_df['Value'], errors='coerce') # numeric value\n",
        "\n",
        "    # mean abundance per phylum/group\n",
        "    mean_abund = loc_df.groupby(['phylum', 'Group'])['Value'].mean().unstack(fill_value=0)\n",
        "\n",
        "    # check if all groups are loaded... to prevent crashing...\n",
        "    if not all(g in mean_abund.columns for g in required_groups):\n",
        "         print(f\"Skipping {loc_name}: Missing required groups {required_groups}. Found: {mean_abund.columns.tolist()}\")\n",
        "         continue\n",
        "\n",
        "    # log2 transform, pseudocount 1\n",
        "    log2_abund = mean_abund.apply(lambda x: np.log2(x + 1))\n",
        "\n",
        "    # log2 fold change ZT22 - ZT10\n",
        "    delta_df = pd.DataFrame({\n",
        "        'WT_change': log2_abund['WT_22'] - log2_abund['WT_10'],\n",
        "        'HET_change': log2_abund['HET_22'] - log2_abund['HET_10']\n",
        "    }, index=log2_abund.index)\n",
        "\n",
        "    processed_data[loc_name] = delta_df\n",
        "    all_phyla.update(delta_df.index)\n",
        "\n",
        "# aligning phyla across all locations + calculate Z-scores\n",
        "aligned_zscored_data = {}\n",
        "global_min_max = {'min': float('inf'), 'max': float('-inf')}\n",
        "\n",
        "sorted_phyla = sorted(list(all_phyla))\n",
        "for loc_name, delta_df in processed_data.items():\n",
        "    # re-index to ensure all phyla are present, fill missing.\n",
        "    aligned_df = delta_df.reindex(sorted_phyla, fill_value=np.nan)\n",
        "    # calculate Z-score per column - WT_change, HET_change across phyla\n",
        "    zscored_df = aligned_df.astype(float).apply(zscore, axis=0, nan_policy='omit')\n",
        "    aligned_zscored_data[loc_name] = zscored_df\n",
        "\n",
        "    # update global min/max for the colouration\n",
        "    current_min, current_max = np.nanmin(zscored_df.values), np.nanmax(zscored_df.values)\n",
        "    if not np.isnan(current_min): global_min_max['min'] = min(global_min_max['min'], current_min)\n",
        "    if not np.isnan(current_max): global_min_max['max'] = max(global_min_max['max'], current_max)\n",
        "\n",
        "# handle no min/max + equal min/max\n",
        "vmin = global_min_max['min'] if np.isfinite(global_min_max['min']) else -1\n",
        "vmax = global_min_max['max'] if np.isfinite(global_min_max['max']) else 1\n",
        "if vmin == vmax: vmin -= 0.5; vmax += 0.5\n",
        "\n",
        "# plotting here\n",
        "plot_order = ['Faecal', 'Ileum', 'Duodenum']\n",
        "valid_plots = [loc for loc in plot_order if loc in aligned_zscored_data]\n",
        "num_plots = len(valid_plots)\n",
        "\n",
        "if num_plots == 0:\n",
        "    print(\"error: no data to plot.\")\n",
        "else:\n",
        "    # fitting figure size\n",
        "    num_phyla = len(sorted_phyla)\n",
        "    num_conditions = 2\n",
        "    cell_width_inches = 0.3\n",
        "    fig_width = max(8, cell_width_inches * num_phyla + 2)\n",
        "    fig_height = cell_width_inches * num_conditions * num_plots + 2.5\n",
        "\n",
        "    fig = plt.figure(figsize=(fig_width, fig_height))\n",
        "    gs = gridspec.GridSpec(num_plots, 1, hspace=0.05, top=0.90, bottom=0.25) # vertical spacing\n",
        "\n",
        "    cmap = sns.diverging_palette(240, 10, as_cmap=True)\n",
        "    cmap = cmap.copy()\n",
        "    cmap.set_bad(color='lightgrey') # for NaN values earlier\n",
        "\n",
        "    ax1 = None # share x-axis\n",
        "\n",
        "    for i, loc_name in enumerate(valid_plots):\n",
        "        ax = fig.add_subplot(gs[i, 0], sharex=ax1)\n",
        "        if i == 0: ax1 = ax # set the axis\n",
        "\n",
        "        data_to_plot = aligned_zscored_data[loc_name].T # transposing for phyla\n",
        "        sns.heatmap(data_to_plot, cmap=cmap, center=0, vmin=vmin, vmax=vmax,\n",
        "                    annot=False, linewidths=.5, ax=ax, cbar=False)\n",
        "\n",
        "        ax.set_ylabel(loc_name, rotation=0, ha='right', va='center', labelpad=15, fontweight='bold')\n",
        "        ax.set_title('')\n",
        "        plt.setp(ax.get_yticklabels(), rotation=0) # y as horizontal\n",
        "\n",
        "        if i == num_plots - 1:\n",
        "            plt.setp(ax.get_xticklabels(), rotation=90, visible=True)\n",
        "            ax.set_xlabel('')\n",
        "        else:\n",
        "            plt.setp(ax.get_xticklabels(), visible=False)\n",
        "            ax.set_xlabel('')\n",
        "\n",
        "        ax.set_aspect('equal', adjustable='box')\n",
        "\n",
        "    # add colour bar\n",
        "    cbar_ax = fig.add_axes([0.15, gs.top + 0.02, 0.7, 0.02]) # no overlap with the plot...\n",
        "    sm = cm.ScalarMappable(cmap=cmap, norm=plt.Normalize(vmin=vmin, vmax=vmax))\n",
        "    sm.set_array([])\n",
        "    cbar = fig.colorbar(sm, cax=cbar_ax, orientation='horizontal')\n",
        "    cbar.ax.xaxis.set_ticks_position('top')\n",
        "    cbar.ax.xaxis.set_label_position('top')\n",
        "\n",
        "    plt.show()\n",
        "\n",
        "#---------------- if crash separate.. or remove\n",
        "\n",
        "import matplotlib\n",
        "matplotlib.use(\"Agg\")\n",
        "import matplotlib.pyplot as plt\n",
        "from PIL import Image\n",
        "\n",
        "fig = plt.gcf()\n",
        "\n",
        "fig.set_size_inches(11, 4)\n",
        "\n",
        "out_png = \"fig7_phylum_change_heatmap.png\"\n",
        "\n",
        "fig.canvas.draw()\n",
        "\n",
        "fig.savefig(out_png,dpi=300,bbox_inches=\"tight\",facecolor=\"white\",format=\"png\")\n",
        "\n",
        "plt.close(fig)\n",
        "\n",
        "# Image.open(out_png).verify()\n",
        "\n"
      ],
      "metadata": {
        "id": "-nGHbSW4wEBS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## FB ratio"
      ],
      "metadata": {
        "id": "LSI7NXklwlFS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Faecal\n",
        "\n",
        "!pip install statannot\n",
        "from statannot import add_stat_annotation\n",
        "\n",
        "\n",
        "# data for Firmicutes + Bacteroidetes\n",
        "firmicutes = merged_df_F[merged_df_F[\"phylum\"] == \"p:Firmicutes\"].copy()\n",
        "bacteroidetes = merged_df_F[merged_df_F[\"phylum\"] == \"p:Bacteroidetes\"].copy()\n",
        "\n",
        "# sum values (for each sample_id)\n",
        "firmicutes_sum = firmicutes.groupby(\"sample_id\")[\"Value\"].sum()\n",
        "bacteroidetes_sum = bacteroidetes.groupby(\"sample_id\")[\"Value\"].sum()\n",
        "\n",
        "# calculate FB ratio\n",
        "fb_ratio = firmicutes_sum / bacteroidetes_sum\n",
        "fb_ratio = fb_ratio.reset_index().rename(columns={0: \"FB_ratio\"})\n",
        "\n",
        "# merge with metadata for plotting\n",
        "fb_ratio = fb_ratio.merge(merged_df_F[[\"sample_id\", \"Group\"]].drop_duplicates(), on=\"sample_id\")\n",
        "\n",
        "# plot FB ratio\n",
        "base = sns.color_palette('pastel', 8)\n",
        "my_palette = dict(zip(grp_order, [base[i] for i in [2, 0, 3, 4]]))\n",
        "\n",
        "plt.figure(figsize=(5, 8))\n",
        "plot_order = [\"WT_10\", \"WT_22\", \"HET_10\",\"HET_22\"]\n",
        "sns.boxplot(x=\"Group\", y=\"Value\", data=fb_ratio, order=plot_order, palette=my_palette)\n",
        "sns.stripplot(x=\"Group\", y=\"Value\", data=fb_ratio, color=\"black\", alpha=0.6, jitter=True, order=plot_order)\n",
        "plt.ylabel(\"Firmicutes-Bacteroidetes Ratio\")\n",
        "plt.xlabel(\"Group\")\n",
        "\n",
        "plt.grid(False)\n",
        "\n",
        "# grab the axes\n",
        "ax = plt.gca()\n",
        "\n",
        "# define pairwise comparisons\n",
        "box_pairs = [(\"WT_10\",\"WT_22\"),(\"HET_10\",\"HET_22\"),(\"WT_10\",\"HET_10\"),(\"WT_22\",\"HET_22\"),]\n",
        "ax = plt.gca()\n",
        "y_max = fb_ratio[\"Value\"].max()\n",
        "h = y_max * 0.05   # height for each bracket\n",
        "\n",
        "for i, (g1, g2) in enumerate(box_pairs):\n",
        "    d1 = fb_ratio.loc[fb_ratio[\"Group\"] == g1, \"Value\"]\n",
        "    d2 = fb_ratio.loc[fb_ratio[\"Group\"] == g2, \"Value\"]\n",
        "    stat, p = mannwhitneyu(d1, d2, alternative=\"two-sided\")\n",
        "\n",
        "    # map p‐value\n",
        "    if   p <= 1e-4: stars = \"****\"\n",
        "    elif p <= 1e-3: stars = \"***\"\n",
        "    elif p <= 1e-2: stars = \"**\"\n",
        "    elif p <= 5e-2: stars = \"*\"\n",
        "    else: stars = None\n",
        "\n",
        "    # if none.\n",
        "    if stars:\n",
        "        x1 = plot_order.index(g1)\n",
        "        x2 = plot_order.index(g2)\n",
        "        y = y_max + h*i\n",
        "\n",
        "        ax.plot([x1,x1,x2,x2],[y,y+h,y+h,y],lw=1.5,c=\"k\") # bracket\n",
        "\n",
        "        ax.text((x1+x2)*0.5,y+h*1.1,stars,ha=\"center\",va=\"bottom\") # stars\n",
        "\n",
        "# save\n",
        "plt.title(\"Faecal\")\n",
        "plt.savefig(\"fig8a_faecal.png\", dpi=300)\n",
        "plt.show()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "leHa9398w16L",
        "outputId": "0b9edddd-1ed1-41f9-8c7c-702bb618662c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: statannot in /usr/local/lib/python3.11/dist-packages (0.2.3)\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-159-2514bd8d09ac>:28: FutureWarning: \n",
            "\n",
            "Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.\n",
            "\n",
            "  sns.boxplot(x=\"Group\", y=\"Value\", data=fb_ratio, order=plot_order, palette=my_palette)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "results = []\n",
        "for g1, g2 in box_pairs:\n",
        "    d1 = fb_ratio.loc[fb_ratio[\"Group\"] == g1, \"Value\"]\n",
        "    d2 = fb_ratio.loc[fb_ratio[\"Group\"] == g2, \"Value\"]\n",
        "    stat, p = mannwhitneyu(d1, d2, alternative=\"two-sided\")\n",
        "    results.append({\"group1\": g1, \"group2\": g2, \"U_stat\": stat, \"p_value\": p})\n",
        "\n",
        "res_df_faecal = pd.DataFrame(results)\n",
        "print(res_df_faecal)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "TRFvwfAaukFe",
        "outputId": "a8e6bdf8-0608-44f6-c4d2-83e37f6c4b99"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "   group1  group2  U_stat   p_value\n",
            "0   WT_10   WT_22    31.0  1.000000\n",
            "1  HET_10  HET_22    32.0  0.606993\n",
            "2   WT_10  HET_10    21.0  0.528671\n",
            "3   WT_22  HET_22    32.0  1.000000\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "faecal_FB = fb_ratio.copy()\n",
        "\n",
        "faecal_FB.to_csv(\"FB_faecal.csv\", index=False)\n",
        "faecal_FB.to_csv(\"FB_faecal.txt\", sep=\"\\t\", index=False)\n",
        "faecal_FB.to_csv(\"FB_faecal.tsv\", sep=\"\\t\", index=False)\n",
        "\n",
        "res_df_faecal.to_csv(\"res_df_faecal_stats.csv\", index=False)\n",
        "res_df_faecal.to_csv(\"res_df_faecal_stats.txt\", sep=\"\\t\", index=False)\n",
        "res_df_faecal.to_csv(\"res_df_faecal_stats.tsv\", sep=\"\\t\", index=False)"
      ],
      "metadata": {
        "id": "1ErxFqVMgjvx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Ileum\n",
        "\n",
        "!pip install statannot\n",
        "from statannot import add_stat_annotation\n",
        "\n",
        "\n",
        "# data for Firmicutes + Bacteroidetes\n",
        "firmicutes = merged_df_I[merged_df_I[\"phylum\"] == \"p:Firmicutes\"].copy()\n",
        "bacteroidetes = merged_df_I[merged_df_I[\"phylum\"] == \"p:Bacteroidetes\"].copy()\n",
        "\n",
        "# sum values (for each sample_id)\n",
        "firmicutes_sum = firmicutes.groupby(\"sample_id\")[\"Value\"].sum()\n",
        "bacteroidetes_sum = bacteroidetes.groupby(\"sample_id\")[\"Value\"].sum()\n",
        "\n",
        "# calculate FB ratio\n",
        "fb_ratio = firmicutes_sum / bacteroidetes_sum\n",
        "fb_ratio = fb_ratio.reset_index().rename(columns={0: \"FB_ratio\"})\n",
        "\n",
        "# merge with metadata for plotting\n",
        "fb_ratio = fb_ratio.merge(merged_df_I[[\"sample_id\", \"Group\"]].drop_duplicates(), on=\"sample_id\")\n",
        "\n",
        "# plot FB ratio\n",
        "base = sns.color_palette('pastel', 8)\n",
        "my_palette = dict(zip(grp_order, [base[i] for i in [2, 0, 3, 4]]))\n",
        "\n",
        "plt.figure(figsize=(5, 8))\n",
        "plot_order = [\"WT_10\", \"WT_22\", \"HET_10\",\"HET_22\"]\n",
        "sns.boxplot(x=\"Group\", y=\"Value\", data=fb_ratio, order=plot_order, palette=my_palette)\n",
        "sns.stripplot(x=\"Group\", y=\"Value\", data=fb_ratio, color=\"black\", alpha=0.6, jitter=True, order=plot_order)\n",
        "plt.ylabel(\"\")\n",
        "plt.xlabel(\"Group\")\n",
        "\n",
        "plt.grid(False)\n",
        "\n",
        "# grab the axes\n",
        "ax = plt.gca()\n",
        "\n",
        "# define pairwise comparisons\n",
        "box_pairs = [(\"WT_10\",\"WT_22\"),(\"HET_10\",\"HET_22\"),(\"WT_10\",\"HET_10\"),(\"WT_22\",\"HET_22\"),]\n",
        "ax    = plt.gca()\n",
        "y_max = fb_ratio[\"Value\"].max()\n",
        "h     = y_max * 0.05   # height for each bracket\n",
        "\n",
        "for i, (g1, g2) in enumerate(box_pairs):\n",
        "    d1 = fb_ratio.loc[fb_ratio[\"Group\"] == g1, \"Value\"]\n",
        "    d2 = fb_ratio.loc[fb_ratio[\"Group\"] == g2, \"Value\"]\n",
        "    stat, p = mannwhitneyu(d1, d2, alternative=\"two-sided\")\n",
        "\n",
        "    # map p‐value\n",
        "    if   p <= 1e-4: stars = \"****\"\n",
        "    elif p <= 1e-3: stars = \"***\"\n",
        "    elif p <= 1e-2: stars = \"**\"\n",
        "    elif p <= 5e-2: stars = \"*\"\n",
        "    else: stars = None\n",
        "\n",
        "    # if none.\n",
        "    if stars:\n",
        "        x1 = plot_order.index(g1)\n",
        "        x2 = plot_order.index(g2)\n",
        "        y = y_max + h*i\n",
        "\n",
        "        ax.plot([x1,x1,x2,x2],[y,y+h,y+h,y],lw=1.5,c=\"k\") # bracket\n",
        "\n",
        "        ax.text((x1+x2)*0.5,y+h*1.1,stars,ha=\"center\",va=\"bottom\") # stars\n",
        "\n",
        "# save\n",
        "plt.title(\"Ileum\")\n",
        "plt.savefig(\"fig8b_ileum.png\", dpi=300)\n",
        "plt.show()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3RhPtJxskqq0",
        "outputId": "346617a1-8302-4f90-d084-c918a1e4ef83"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: statannot in /usr/local/lib/python3.11/dist-packages (0.2.3)\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-162-2bdcae364326>:28: FutureWarning: \n",
            "\n",
            "Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.\n",
            "\n",
            "  sns.boxplot(x=\"Group\", y=\"Value\", data=fb_ratio, order=plot_order, palette=my_palette)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "results = []\n",
        "for g1, g2 in box_pairs:\n",
        "    d1 = fb_ratio.loc[fb_ratio[\"Group\"] == g1, \"Value\"]\n",
        "    d2 = fb_ratio.loc[fb_ratio[\"Group\"] == g2, \"Value\"]\n",
        "    stat, p = mannwhitneyu(d1, d2, alternative=\"two-sided\")\n",
        "    results.append({\"group1\": g1, \"group2\": g2, \"U_stat\": stat, \"p_value\": p})\n",
        "\n",
        "res_df_ileum = pd.DataFrame(results)\n",
        "print(res_df_ileum)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_4bp3A1rw33x",
        "outputId": "b8348bb1-a15e-4f37-faf1-ee2d92f6b25a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "   group1  group2  U_stat   p_value\n",
            "0   WT_10   WT_22    19.0  0.572761\n",
            "1  HET_10  HET_22    12.0  1.000000\n",
            "2   WT_10  HET_10    25.0  0.523699\n",
            "3   WT_22  HET_22    17.0  0.792208\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "ileum_FB = fb_ratio.copy()\n",
        "\n",
        "ileum_FB.to_csv(\"FB_ileum.csv\", index=False)\n",
        "ileum_FB.to_csv(\"FB_ileum.txt\", sep=\"\\t\", index=False)\n",
        "ileum_FB.to_csv(\"FB_ileum.tsv\", sep=\"\\t\", index=False)\n",
        "\n",
        "res_df_ileum.to_csv(\"res_df_ileum_stats.csv\", index=False)\n",
        "res_df_ileum.to_csv(\"res_df_ileum_stats.txt\", sep=\"\\t\", index=False)\n",
        "res_df_ileum.to_csv(\"res_df_ileum_stats.tsv\", sep=\"\\t\", index=False)"
      ],
      "metadata": {
        "id": "q7u6yURsy_DU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Duodenum\n",
        "\n",
        "!pip install statannot\n",
        "from statannot import add_stat_annotation\n",
        "\n",
        "\n",
        "# data for Firmicutes + Bacteroidetes\n",
        "firmicutes = merged_df_ST[merged_df_ST[\"phylum\"] == \"p:Firmicutes\"].copy()\n",
        "bacteroidetes = merged_df_ST[merged_df_ST[\"phylum\"] == \"p:Bacteroidetes\"].copy()\n",
        "\n",
        "# sum values (for each sample_id)\n",
        "firmicutes_sum = firmicutes.groupby(\"sample_id\")[\"Value\"].sum()\n",
        "bacteroidetes_sum = bacteroidetes.groupby(\"sample_id\")[\"Value\"].sum()\n",
        "\n",
        "# calculate FB ratio\n",
        "fb_ratio = firmicutes_sum / bacteroidetes_sum\n",
        "fb_ratio = fb_ratio.reset_index().rename(columns={0: \"FB_ratio\"})\n",
        "\n",
        "# merge with metadata for plotting\n",
        "fb_ratio = fb_ratio.merge(merged_df_ST[[\"sample_id\", \"Group\"]].drop_duplicates(), on=\"sample_id\")\n",
        "\n",
        "# plot FB ratio\n",
        "base = sns.color_palette('pastel', 8)\n",
        "my_palette = dict(zip(grp_order, [base[i] for i in [2, 0, 3, 4]]))\n",
        "\n",
        "plt.figure(figsize=(5, 8))\n",
        "plot_order = [\"WT_10\", \"WT_22\", \"HET_10\",\"HET_22\"]\n",
        "sns.boxplot(x=\"Group\", y=\"Value\", data=fb_ratio, order=plot_order, palette=my_palette)\n",
        "sns.stripplot(x=\"Group\", y=\"Value\", data=fb_ratio, color=\"black\", alpha=0.6, jitter=True, order=plot_order)\n",
        "plt.ylabel(\"\")\n",
        "plt.xlabel(\"Group\")\n",
        "\n",
        "plt.grid(False)\n",
        "\n",
        "# grab the axes\n",
        "ax = plt.gca()\n",
        "\n",
        "# define pairwise comparisons\n",
        "box_pairs = [(\"WT_10\",\"WT_22\"),(\"HET_10\",\"HET_22\"),(\"WT_10\",\"HET_10\"),(\"WT_22\",\"HET_22\"),]\n",
        "ax    = plt.gca()\n",
        "y_max = fb_ratio[\"Value\"].max()\n",
        "h     = y_max * 0.05   # height for each bracket\n",
        "\n",
        "for i, (g1, g2) in enumerate(box_pairs):\n",
        "    d1 = fb_ratio.loc[fb_ratio[\"Group\"] == g1, \"Value\"]\n",
        "    d2 = fb_ratio.loc[fb_ratio[\"Group\"] == g2, \"Value\"]\n",
        "    stat, p = mannwhitneyu(d1, d2, alternative=\"two-sided\")\n",
        "\n",
        "    # map p‐value\n",
        "    if   p <= 1e-4: stars = \"****\"\n",
        "    elif p <= 1e-3: stars = \"***\"\n",
        "    elif p <= 1e-2: stars = \"**\"\n",
        "    elif p <= 5e-2: stars = \"*\"\n",
        "    else: stars = None\n",
        "\n",
        "    # if none.\n",
        "    if stars:\n",
        "        x1 = plot_order.index(g1)\n",
        "        x2 = plot_order.index(g2)\n",
        "        y = y_max + h*i\n",
        "\n",
        "        ax.plot([x1,x1,x2,x2],[y,y+h,y+h,y],lw=1.5,c=\"k\") # bracket\n",
        "\n",
        "        ax.text((x1+x2)*0.5,y+h*1.1,stars,ha=\"center\",va=\"bottom\") # stars\n",
        "\n",
        "# save\n",
        "plt.title(\"Duodenum\")\n",
        "plt.savefig(\"fig8c_duodenum.png\", dpi=300)\n",
        "plt.show()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5y6o3sAyzTzG",
        "outputId": "b3a22908-ff0c-46ac-9075-c2f1467648fa"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: statannot in /usr/local/lib/python3.11/dist-packages (0.2.3)\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-165-abbd241442b3>:28: FutureWarning: \n",
            "\n",
            "Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.\n",
            "\n",
            "  sns.boxplot(x=\"Group\", y=\"Value\", data=fb_ratio, order=plot_order, palette=my_palette)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "results = []\n",
        "for g1, g2 in box_pairs:\n",
        "    d1 = fb_ratio.loc[fb_ratio[\"Group\"] == g1, \"Value\"]\n",
        "    d2 = fb_ratio.loc[fb_ratio[\"Group\"] == g2, \"Value\"]\n",
        "    stat, p = mannwhitneyu(d1, d2, alternative=\"two-sided\")\n",
        "    results.append({\"group1\": g1, \"group2\": g2, \"U_stat\": stat, \"p_value\": p})\n",
        "\n",
        "res_df_duo = pd.DataFrame(results)\n",
        "print(res_df_duo)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sWQAU_YGxL3g",
        "outputId": "0e3436a8-3977-47db-aa19-fb5d8d1c794a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "   group1  group2  U_stat   p_value\n",
            "0   WT_10   WT_22     5.0  0.571429\n",
            "1  HET_10  HET_22     8.0  1.000000\n",
            "2   WT_10  HET_10    12.0  0.730159\n",
            "3   WT_22  HET_22    10.0  0.228571\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "duo_FB = fb_ratio.copy()\n",
        "\n",
        "duo_FB.to_csv(\"duo_FB.csv\", index=False)\n",
        "duo_FB.to_csv(\"duo_FB.txt\", sep=\"\\t\", index=False)\n",
        "duo_FB.to_csv(\"duo_FB.tsv\", sep=\"\\t\", index=False)\n",
        "\n",
        "res_df_duo.to_csv(\"res_df_duo_stats.csv\", index=False)\n",
        "res_df_duo.to_csv(\"res_df_duo_stats.txt\", sep=\"\\t\", index=False)\n",
        "res_df_duo.to_csv(\"res_df_duo_stats.tsv\", sep=\"\\t\", index=False)"
      ],
      "metadata": {
        "id": "C1caEZ-Qg1lr"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## volcano plot + log-fold change rank"
      ],
      "metadata": {
        "id": "5KncaFuPwmvN"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from skbio.stats.composition import clr, ilr, centralize, dirmult_ttest, ancom, multi_replace\n",
        "\n",
        "feature_table =pd.read_csv('rare_table.txt', sep='\\t', index_col=0)\n",
        "abundance_data = feature_table.apply(multi_replace)\n",
        "\n",
        "abundance_data = abundance_data.T"
      ],
      "metadata": {
        "id": "hzYpu8V2wzAm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# check if indexes match.\n",
        "abundance_data.index\n",
        "metadata.index"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KtowFgJdxi7k",
        "outputId": "f74c6b31-c696-495b-a785-f2c127818972"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Index(['F1', 'F2', 'F3', 'F4', 'F5', 'F6', 'F7', 'F8', 'F9', 'F10', 'F11',\n",
              "       'F12', 'F13', 'F14', 'F15', 'F16', 'F17', 'F18', 'F19', 'F20', 'F21',\n",
              "       'F22', 'IN1', 'IN2', 'IN3', 'IN4', 'IN5', 'IN6', 'IN7', 'IN8', 'IN9',\n",
              "       'IN10', 'IN11', 'IN12', 'IN13', 'IN14', 'IN15', 'IN16', 'IN18', 'IN19',\n",
              "       'IN20', 'IN21', 'IN22', 'IN23', 'IN24', 'IN25', 'ST1', 'ST2', 'ST3',\n",
              "       'ST4', 'ST5', 'ST6', 'ST7', 'ST8', 'ST9', 'ST10', 'ST11', 'ST12',\n",
              "       'ST13', 'ST14', 'ST15', 'F23', 'F24', 'F25', 'F26', 'F27', 'F28', 'F29',\n",
              "       'F30', 'F31', 'ST16'],\n",
              "      dtype='object', name='Sample_id')"
            ]
          },
          "metadata": {},
          "execution_count": 169
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "res_volcano = dirmult_ttest(abundance_data, metadata['Sample Name in Report'],\n",
        "                    treatment='HET22', reference='WT22',\n",
        "                    draws=3)\n",
        "res_volcano"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "eQT6wjcvxomZ",
        "outputId": "2febabe5-59a3-46db-f318-7903f3d72f97"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "          T statistic        df  Log2(FC)    CI(2.5)   CI(97.5)    pvalue  \\\n",
              "zotu1        0.262740  4.987283  0.515249  -5.827937   7.868676  0.532617   \n",
              "zotu10       0.565706  2.569646  1.458013  -6.781109  11.865544  0.457068   \n",
              "zotu100     -0.369023  3.991205 -0.436601  -7.908578   9.755407  0.631129   \n",
              "zotu1000    -0.509524  3.052756 -1.439433 -11.425405  10.102392  0.621530   \n",
              "zotu1001     1.277115  2.184459  1.550212  -8.947060  14.712738  0.259827   \n",
              "...               ...       ...       ...        ...        ...       ...   \n",
              "zotu995      0.915277  4.996289  0.920521 -11.648676   7.254651  0.307040   \n",
              "zotu996     -0.285127  4.424248 -1.540147 -15.546755  10.882485  0.351359   \n",
              "zotu997     -0.857689  3.375127 -2.338365 -15.580334  16.409712  0.449035   \n",
              "zotu998     -1.142819  4.336526 -2.345529 -10.138342   9.644044  0.359760   \n",
              "zotu999      0.273595  4.789158  0.332504  -7.737119   8.112454  0.591607   \n",
              "\n",
              "          qvalue  Reject null hypothesis  \n",
              "zotu1        1.0                   False  \n",
              "zotu10       1.0                   False  \n",
              "zotu100      1.0                   False  \n",
              "zotu1000     1.0                   False  \n",
              "zotu1001     1.0                   False  \n",
              "...          ...                     ...  \n",
              "zotu995      1.0                   False  \n",
              "zotu996      1.0                   False  \n",
              "zotu997      1.0                   False  \n",
              "zotu998      1.0                   False  \n",
              "zotu999      1.0                   False  \n",
              "\n",
              "[1503 rows x 8 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-ad468c1a-0f96-43ee-9ddd-44a39780fe4a\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>T statistic</th>\n",
              "      <th>df</th>\n",
              "      <th>Log2(FC)</th>\n",
              "      <th>CI(2.5)</th>\n",
              "      <th>CI(97.5)</th>\n",
              "      <th>pvalue</th>\n",
              "      <th>qvalue</th>\n",
              "      <th>Reject null hypothesis</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>zotu1</th>\n",
              "      <td>0.262740</td>\n",
              "      <td>4.987283</td>\n",
              "      <td>0.515249</td>\n",
              "      <td>-5.827937</td>\n",
              "      <td>7.868676</td>\n",
              "      <td>0.532617</td>\n",
              "      <td>1.0</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>zotu10</th>\n",
              "      <td>0.565706</td>\n",
              "      <td>2.569646</td>\n",
              "      <td>1.458013</td>\n",
              "      <td>-6.781109</td>\n",
              "      <td>11.865544</td>\n",
              "      <td>0.457068</td>\n",
              "      <td>1.0</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>zotu100</th>\n",
              "      <td>-0.369023</td>\n",
              "      <td>3.991205</td>\n",
              "      <td>-0.436601</td>\n",
              "      <td>-7.908578</td>\n",
              "      <td>9.755407</td>\n",
              "      <td>0.631129</td>\n",
              "      <td>1.0</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>zotu1000</th>\n",
              "      <td>-0.509524</td>\n",
              "      <td>3.052756</td>\n",
              "      <td>-1.439433</td>\n",
              "      <td>-11.425405</td>\n",
              "      <td>10.102392</td>\n",
              "      <td>0.621530</td>\n",
              "      <td>1.0</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>zotu1001</th>\n",
              "      <td>1.277115</td>\n",
              "      <td>2.184459</td>\n",
              "      <td>1.550212</td>\n",
              "      <td>-8.947060</td>\n",
              "      <td>14.712738</td>\n",
              "      <td>0.259827</td>\n",
              "      <td>1.0</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>zotu995</th>\n",
              "      <td>0.915277</td>\n",
              "      <td>4.996289</td>\n",
              "      <td>0.920521</td>\n",
              "      <td>-11.648676</td>\n",
              "      <td>7.254651</td>\n",
              "      <td>0.307040</td>\n",
              "      <td>1.0</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>zotu996</th>\n",
              "      <td>-0.285127</td>\n",
              "      <td>4.424248</td>\n",
              "      <td>-1.540147</td>\n",
              "      <td>-15.546755</td>\n",
              "      <td>10.882485</td>\n",
              "      <td>0.351359</td>\n",
              "      <td>1.0</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>zotu997</th>\n",
              "      <td>-0.857689</td>\n",
              "      <td>3.375127</td>\n",
              "      <td>-2.338365</td>\n",
              "      <td>-15.580334</td>\n",
              "      <td>16.409712</td>\n",
              "      <td>0.449035</td>\n",
              "      <td>1.0</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>zotu998</th>\n",
              "      <td>-1.142819</td>\n",
              "      <td>4.336526</td>\n",
              "      <td>-2.345529</td>\n",
              "      <td>-10.138342</td>\n",
              "      <td>9.644044</td>\n",
              "      <td>0.359760</td>\n",
              "      <td>1.0</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>zotu999</th>\n",
              "      <td>0.273595</td>\n",
              "      <td>4.789158</td>\n",
              "      <td>0.332504</td>\n",
              "      <td>-7.737119</td>\n",
              "      <td>8.112454</td>\n",
              "      <td>0.591607</td>\n",
              "      <td>1.0</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1503 rows × 8 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-ad468c1a-0f96-43ee-9ddd-44a39780fe4a')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-ad468c1a-0f96-43ee-9ddd-44a39780fe4a button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-ad468c1a-0f96-43ee-9ddd-44a39780fe4a');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-fc7ccd76-9bdb-47c2-87a6-9c61a16e3978\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-fc7ccd76-9bdb-47c2-87a6-9c61a16e3978')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-fc7ccd76-9bdb-47c2-87a6-9c61a16e3978 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "  <div id=\"id_9adaaf92-621e-4d6d-9057-800ee36b4abd\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('res_volcano')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_9adaaf92-621e-4d6d-9057-800ee36b4abd button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('res_volcano');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "res_volcano",
              "summary": "{\n  \"name\": \"res_volcano\",\n  \"rows\": 1503,\n  \"fields\": [\n    {\n      \"column\": \"T statistic\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.7297524020908153,\n        \"min\": -2.627021803773277,\n        \"max\": 2.86270065363605,\n        \"num_unique_values\": 1503,\n        \"samples\": [\n          -0.43625489175978416,\n          0.5486182164564375,\n          0.20504667874890767\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"df\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.9104984256146101,\n        \"min\": 2.0045042938798616,\n        \"max\": 4.999998961280848,\n        \"num_unique_values\": 1503,\n        \"samples\": [\n          4.716333664112695,\n          4.295052499288915,\n          2.832192572531925\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Log2(FC)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 1.409172444090742,\n        \"min\": -4.90778282525552,\n        \"max\": 4.45343527570638,\n        \"num_unique_values\": 1503,\n        \"samples\": [\n          -1.4102528543064627,\n          2.2669563463454865,\n          0.03542882133550634\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"CI(2.5)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 3.6629045983581934,\n        \"min\": -31.689314174305153,\n        \"max\": -1.7742902460320151,\n        \"num_unique_values\": 1503,\n        \"samples\": [\n          -12.374053413321848,\n          -8.660321892841525,\n          -16.348222306668845\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"CI(97.5)\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 5.454006178614004,\n        \"min\": 1.3301127181730585,\n        \"max\": 44.54139670891527,\n        \"num_unique_values\": 1503,\n        \"samples\": [\n          11.137978718638857,\n          16.25528783667868,\n          13.407205150335002\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"pvalue\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.15670751439003638,\n        \"min\": 0.04993889803643189,\n        \"max\": 0.9342163068655126,\n        \"num_unique_values\": 1503,\n        \"samples\": [\n          0.6136619217076877,\n          0.4439955568021281,\n          0.45925914479095575\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"qvalue\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 1.0,\n        \"max\": 1.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          1.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Reject null hypothesis\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          false\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 170
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# volcano plot\n",
        "# Compute -log10(pvalue)\n",
        "import numpy as np\n",
        "res_volcano['neg_log10_pval'] = -np.log10(res_volcano['pvalue'])\n",
        "\n",
        "# Plot\n",
        "plt.figure(figsize=(8, 6))\n",
        "sns.scatterplot(\n",
        "    data=res_volcano,\n",
        "    x=\"Log2(FC)\",\n",
        "    y=\"neg_log10_pval\",\n",
        "    hue=\"Reject null hypothesis\",\n",
        "    palette={True: \"red\", False: \"gray\"},\n",
        "    edgecolor=\"black\"\n",
        ")\n",
        "\n",
        "plt.axhline(y=-np.log10(0.05), color='blue', linestyle='--', linewidth=1)\n",
        "plt.axvline(x=1, color='green', linestyle='--', linewidth=1)\n",
        "plt.axvline(x=-1, color='green', linestyle='--', linewidth=1)\n",
        "\n",
        "plt.title(\"Volcano Plot\")\n",
        "plt.xlabel(\"Log2 Fold Change\")\n",
        "plt.ylabel(\"-log10(p-value)\")\n",
        "plt.legend(title=\"Significant\")\n",
        "plt.tight_layout()\n",
        "plt.show()\n",
        "\n",
        "plt.savefig(\"fig9.png\")"
      ],
      "metadata": {
        "id": "UTGBGyhEyBCa"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# extract rank information by sorting by log-fold-change\n",
        "ttest_res = (res_volcano\n",
        "             .sort_values('Log2(FC)')\n",
        "             .reset_index(names='featureid')\n",
        "             .reset_index(names='rank'))\n",
        "\n",
        "sns.scatterplot(data=ttest_res, x='rank', y='Log2(FC)', s=5, color='b', hue='Reject null hypothesis')\n",
        "sns.scatterplot(data=ttest_res, x='rank', y='CI(2.5)', s=5, color='b', hue='Reject null hypothesis', legend=False)\n",
        "sns.scatterplot(data=ttest_res, x='rank', y='CI(97.5)', s=5, color='b', hue='Reject null hypothesis', legend=False)\n",
        "plt.hlines(0, xmin=0, xmax=len(res_volcano), color='r')\n",
        "\n",
        "plt.savefig(\"fig10.png\")"
      ],
      "metadata": {
        "id": "sWOvE2NvyHoQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "data_to_plot"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 161
        },
        "id": "dD4dZuWsMDXV",
        "outputId": "dc7fd0da-fdf0-4d94-be53-f68a31119f48"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "phylum      p:Acidobacteria  p:Actinobacteria  p:Armatimonadetes  \\\n",
              "WT_change         -2.292857         -0.779832           0.204168   \n",
              "HET_change         0.045727          0.439009           0.045727   \n",
              "\n",
              "phylum      p:Bacteroidetes  p:Campilobacterota  \\\n",
              "WT_change          0.193259            0.202890   \n",
              "HET_change         0.142811            0.962507   \n",
              "\n",
              "phylum      p:Candidatus_Saccharibacteria  p:Chlamydiae  p:Chloroflexi  \\\n",
              "WT_change                        0.632730      0.204168       0.204168   \n",
              "HET_change                       0.082813     -1.479929       0.045727   \n",
              "\n",
              "phylum      p:Cyanobacteria/Chloroplast  p:Deferribacteres  ...  \\\n",
              "WT_change                      1.840424           1.021868  ...   \n",
              "HET_change                    -0.404012           1.426599  ...   \n",
              "\n",
              "phylum      p:Latescibacteria  p:Nitrospirae  p:Parcubacteria  \\\n",
              "WT_change            0.204168      -0.668227         2.015072   \n",
              "HET_change           0.045727       0.045727         0.045727   \n",
              "\n",
              "phylum      p:Planctomycetes  p:Plantae  p:Proteobacteria  p:Rhodothermaeota  \\\n",
              "WT_change          -0.696507  -0.817592          0.451049           0.204168   \n",
              "HET_change          0.045727   2.700116         -0.714525           0.045727   \n",
              "\n",
              "phylum         p:SR1  p:Tenericutes  p:Verrucomicrobia  \n",
              "WT_change   0.204168      -1.961885          -0.856531  \n",
              "HET_change  0.045727      -0.897911           0.045727  \n",
              "\n",
              "[2 rows x 24 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-26af6761-4939-4d56-b28b-7fa5e2b90f52\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>phylum</th>\n",
              "      <th>p:Acidobacteria</th>\n",
              "      <th>p:Actinobacteria</th>\n",
              "      <th>p:Armatimonadetes</th>\n",
              "      <th>p:Bacteroidetes</th>\n",
              "      <th>p:Campilobacterota</th>\n",
              "      <th>p:Candidatus_Saccharibacteria</th>\n",
              "      <th>p:Chlamydiae</th>\n",
              "      <th>p:Chloroflexi</th>\n",
              "      <th>p:Cyanobacteria/Chloroplast</th>\n",
              "      <th>p:Deferribacteres</th>\n",
              "      <th>...</th>\n",
              "      <th>p:Latescibacteria</th>\n",
              "      <th>p:Nitrospirae</th>\n",
              "      <th>p:Parcubacteria</th>\n",
              "      <th>p:Planctomycetes</th>\n",
              "      <th>p:Plantae</th>\n",
              "      <th>p:Proteobacteria</th>\n",
              "      <th>p:Rhodothermaeota</th>\n",
              "      <th>p:SR1</th>\n",
              "      <th>p:Tenericutes</th>\n",
              "      <th>p:Verrucomicrobia</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>WT_change</th>\n",
              "      <td>-2.292857</td>\n",
              "      <td>-0.779832</td>\n",
              "      <td>0.204168</td>\n",
              "      <td>0.193259</td>\n",
              "      <td>0.202890</td>\n",
              "      <td>0.632730</td>\n",
              "      <td>0.204168</td>\n",
              "      <td>0.204168</td>\n",
              "      <td>1.840424</td>\n",
              "      <td>1.021868</td>\n",
              "      <td>...</td>\n",
              "      <td>0.204168</td>\n",
              "      <td>-0.668227</td>\n",
              "      <td>2.015072</td>\n",
              "      <td>-0.696507</td>\n",
              "      <td>-0.817592</td>\n",
              "      <td>0.451049</td>\n",
              "      <td>0.204168</td>\n",
              "      <td>0.204168</td>\n",
              "      <td>-1.961885</td>\n",
              "      <td>-0.856531</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>HET_change</th>\n",
              "      <td>0.045727</td>\n",
              "      <td>0.439009</td>\n",
              "      <td>0.045727</td>\n",
              "      <td>0.142811</td>\n",
              "      <td>0.962507</td>\n",
              "      <td>0.082813</td>\n",
              "      <td>-1.479929</td>\n",
              "      <td>0.045727</td>\n",
              "      <td>-0.404012</td>\n",
              "      <td>1.426599</td>\n",
              "      <td>...</td>\n",
              "      <td>0.045727</td>\n",
              "      <td>0.045727</td>\n",
              "      <td>0.045727</td>\n",
              "      <td>0.045727</td>\n",
              "      <td>2.700116</td>\n",
              "      <td>-0.714525</td>\n",
              "      <td>0.045727</td>\n",
              "      <td>0.045727</td>\n",
              "      <td>-0.897911</td>\n",
              "      <td>0.045727</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>2 rows × 24 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-26af6761-4939-4d56-b28b-7fa5e2b90f52')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-26af6761-4939-4d56-b28b-7fa5e2b90f52 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-26af6761-4939-4d56-b28b-7fa5e2b90f52');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-b804e8d8-f17e-420e-a26a-ad6a4d2ee4f1\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-b804e8d8-f17e-420e-a26a-ad6a4d2ee4f1')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-b804e8d8-f17e-420e-a26a-ad6a4d2ee4f1 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "  <div id=\"id_0ab2bfbd-babd-4cba-9186-fbbeb175daca\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('data_to_plot')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_0ab2bfbd-babd-4cba-9186-fbbeb175daca button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('data_to_plot');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "data_to_plot"
            }
          },
          "metadata": {},
          "execution_count": 173
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# unused"
      ],
      "metadata": {
        "id": "p-ZS1W2HMjVO"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## per-genus plots"
      ],
      "metadata": {
        "id": "nydIqCYzMkw_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "# Faecal, genera\n",
        "# Create a FacetGrid for boxplots\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "\n",
        "g_genera = sns.FacetGrid(merged_df_genera_F, col=\"genus\", col_wrap=3, height=4, sharey=False)\n",
        "\n",
        "# Map boxplot to each facet\n",
        "g_genera.map_dataframe(sns.boxplot, x=\"Group\", y=\"Value\", showfliers=False, order=sorted(merged_df_genera_F[\"Group\"].unique()))\n",
        "g_genera.map_dataframe(sns.swarmplot, x=\"Group\", y=\"Value\", color=\".25\", order=sorted(merged_df_genera_F[\"Group\"].unique()))\n",
        "\n",
        "g_genera.set_axis_labels(\"Group\", \"Counts\")\n",
        "g_genera.set_titles(\"{col_name}\")\n",
        "\n",
        "g_genera.tight_layout()\n",
        "\n",
        "plt.show()\n",
        "\n",
        "plt.savefig('fig6a_faecal_genera.png')\n",
        "\"\"\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 87
        },
        "id": "-7RmR2JLMwdZ",
        "outputId": "9496baa7-6ffc-4f8e-b2e7-807c5e708ffc"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'\\n# Faecal, genera\\n# Create a FacetGrid for boxplots\\nimport seaborn as sns\\nimport matplotlib.pyplot as plt\\n\\n\\ng_genera = sns.FacetGrid(merged_df_genera_F, col=\"genus\", col_wrap=3, height=4, sharey=False)\\n\\n# Map boxplot to each facet\\ng_genera.map_dataframe(sns.boxplot, x=\"Group\", y=\"Value\", showfliers=False, order=sorted(merged_df_genera_F[\"Group\"].unique()))\\ng_genera.map_dataframe(sns.swarmplot, x=\"Group\", y=\"Value\", color=\".25\", order=sorted(merged_df_genera_F[\"Group\"].unique()))\\n\\ng_genera.set_axis_labels(\"Group\", \"Counts\")\\ng_genera.set_titles(\"{col_name}\")\\n\\ng_genera.tight_layout()\\n\\nplt.show()\\n\\nplt.savefig(\\'fig6a_faecal_genera.png\\')\\n'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 174
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "# Ileum, genera\n",
        "# Create a FacetGrid for boxplots\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "\n",
        "g_genera = sns.FacetGrid(merged_df_genera_I, col=\"genus\", col_wrap=3, height=4, sharey=False)\n",
        "\n",
        "# Map boxplot to each facet\n",
        "g_genera.map_dataframe(sns.boxplot, x=\"Group\", y=\"Value\", showfliers=False, order=sorted(merged_df_genera_I[\"Group\"].unique()))\n",
        "g_genera.map_dataframe(sns.swarmplot, x=\"Group\", y=\"Value\", color=\".25\", order=sorted(merged_df_genera_I[\"Group\"].unique()))\n",
        "\n",
        "g_genera.set_axis_labels(\"Group\", \"Counts\")\n",
        "g_genera.set_titles(\"{col_name}\")\n",
        "\n",
        "g_genera.tight_layout()\n",
        "\n",
        "plt.show()\n",
        "\n",
        "plt.savefig('fig6b_ileum_genera.png')\n",
        "\"\"\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 87
        },
        "id": "80i_siRmMxUu",
        "outputId": "cf2bb9fb-351e-4512-fc22-206007f4adf9"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'\\n# Ileum, genera\\n# Create a FacetGrid for boxplots\\nimport seaborn as sns\\nimport matplotlib.pyplot as plt\\n\\n\\ng_genera = sns.FacetGrid(merged_df_genera_I, col=\"genus\", col_wrap=3, height=4, sharey=False)\\n\\n# Map boxplot to each facet\\ng_genera.map_dataframe(sns.boxplot, x=\"Group\", y=\"Value\", showfliers=False, order=sorted(merged_df_genera_I[\"Group\"].unique()))\\ng_genera.map_dataframe(sns.swarmplot, x=\"Group\", y=\"Value\", color=\".25\", order=sorted(merged_df_genera_I[\"Group\"].unique()))\\n\\ng_genera.set_axis_labels(\"Group\", \"Counts\")\\ng_genera.set_titles(\"{col_name}\")\\n\\ng_genera.tight_layout()\\n\\nplt.show()\\n\\nplt.savefig(\\'fig6b_ileum_genera.png\\')\\n'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 175
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "# Duo, genera\n",
        "# Create a FacetGrid for boxplots\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "\n",
        "g_genera = sns.FacetGrid(merged_df_genera_ST, col=\"genus\", col_wrap=3, height=4, sharey=False)\n",
        "\n",
        "# Map boxplot to each facet\n",
        "g_genera.map_dataframe(sns.boxplot, x=\"Group\", y=\"Value\", showfliers=False, order=sorted(merged_df_genera_ST[\"Group\"].unique()))\n",
        "g_genera.map_dataframe(sns.swarmplot, x=\"Group\", y=\"Value\", color=\".25\", order=sorted(merged_df_genera_ST[\"Group\"].unique()))\n",
        "\n",
        "g_genera.set_axis_labels(\"Group\", \"Counts\")\n",
        "g_genera.set_titles(\"{col_name}\")\n",
        "\n",
        "g_genera.tight_layout()\n",
        "\n",
        "plt.show()\n",
        "\n",
        "plt.savefig('fig6c_duodenum_genera.png')\n",
        "\"\"\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 87
        },
        "id": "pD_g9dj8Mz7r",
        "outputId": "b5a38a2f-5f39-4031-bd4f-caa114f7eb26"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'\\n# Duo, genera\\n# Create a FacetGrid for boxplots\\nimport seaborn as sns\\nimport matplotlib.pyplot as plt\\n\\n\\ng_genera = sns.FacetGrid(merged_df_genera_ST, col=\"genus\", col_wrap=3, height=4, sharey=False)\\n\\n# Map boxplot to each facet\\ng_genera.map_dataframe(sns.boxplot, x=\"Group\", y=\"Value\", showfliers=False, order=sorted(merged_df_genera_ST[\"Group\"].unique()))\\ng_genera.map_dataframe(sns.swarmplot, x=\"Group\", y=\"Value\", color=\".25\", order=sorted(merged_df_genera_ST[\"Group\"].unique()))\\n\\ng_genera.set_axis_labels(\"Group\", \"Counts\")\\ng_genera.set_titles(\"{col_name}\")\\n\\ng_genera.tight_layout()\\n\\nplt.show()\\n\\nplt.savefig(\\'fig6c_duodenum_genera.png\\')\\n'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 176
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### heatmap (genera)"
      ],
      "metadata": {
        "id": "RgK7ddurMoXa"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "from scipy.stats import zscore\n",
        "import matplotlib.pyplot as plt\n",
        "import matplotlib.gridspec as gridspec\n",
        "import matplotlib.cm as cm\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "\n",
        "# genus copy\n",
        "locations_data_genera = {\n",
        "    'Faecal': merged_df_genera_F.copy() if 'merged_df_genera_F' in locals() and merged_df_genera_F is not None else None,\n",
        "    'Ileum': merged_df_genera_I.copy() if 'merged_df_genera_I' in locals() and merged_df_genera_I is not None else None,\n",
        "    'Duodenum': merged_df_genera_ST.copy() if 'merged_df_genera_ST' in locals() and merged_df_genera_ST is not None else None\n",
        "}\n",
        "\n",
        "processed_data_genera = {}\n",
        "all_genera = set()\n",
        "\n",
        "wt_groups = ['WT_10', 'WT_22']\n",
        "het_groups = ['HET_10', 'HET_22']\n",
        "required_groups = wt_groups + het_groups\n",
        "\n",
        "for loc_name, loc_df_genera in locations_data_genera.items():\n",
        "    if loc_df_genera is None or loc_df_genera.empty:\n",
        "        print(f\"Skipping {loc_name} (Genus): No data.\")\n",
        "        continue\n",
        "\n",
        "    # check\n",
        "    loc_df_genera.loc[:, 'Value'] = pd.to_numeric(loc_df_genera['Value'], errors='coerce')\n",
        "\n",
        "    mean_abund_genera = loc_df_genera.groupby(['genus', 'Group'])['Value'].mean().unstack(fill_value=0)\n",
        "\n",
        "    # check\n",
        "    if not all(g in mean_abund_genera.columns for g in required_groups):\n",
        "        print(f\"Skipping {loc_name} (Genus): Missing required groups {required_groups}. Found: {mean_abund_genera.columns.tolist()}\")\n",
        "        continue\n",
        "\n",
        "    # log2 transform. pseudocount 1\n",
        "    log2_abund_genera = mean_abund_genera.apply(lambda x: np.log2(x + 1))\n",
        "\n",
        "    if all(g in log2_abund_genera.columns for g in wt_groups) and all(g in log2_abund_genera.columns for g in het_groups):\n",
        "        delta_df_genera = pd.DataFrame({\n",
        "            'WT_change': log2_abund_genera['WT_22'] - log2_abund_genera['WT_10'],\n",
        "            'HET_change': log2_abund_genera['HET_22'] - log2_abund_genera['HET_10']\n",
        "        }, index=log2_abund_genera.index) # genus names\n",
        "\n",
        "        processed_data_genera[loc_name] = delta_df_genera\n",
        "        all_genera.update(delta_df_genera.index)\n",
        "    else:\n",
        "         print(f\"Skipping delta calculation for {loc_name} (Genus): Missing one or more required time point groups in log2 data.\")\n",
        "\n",
        "\n",
        "# aligning genera. again check no overwriting BEFORE running please please please.\n",
        "aligned_zscored_data_genera = {}\n",
        "global_min_max_genera = {'min': float('inf'), 'max': float('-inf')}\n",
        "\n",
        "if not all_genera:\n",
        "    print(\"no genera found after processing\")\n",
        "else:\n",
        "    sorted_genera = sorted(list(all_genera))\n",
        "    print(f\"processing {len(sorted_genera)} genera..\") # check again\n",
        "\n",
        "    for loc_name, delta_df_genera in processed_data_genera.items():\n",
        "        aligned_df_genera = delta_df_genera.reindex(sorted_genera, fill_value=np.nan)\n",
        "\n",
        "        zscored_df_genera = aligned_df_genera.astype(float).apply(zscore, axis=0, nan_policy='omit')\n",
        "        aligned_zscored_data_genera[loc_name] = zscored_df_genera\n",
        "\n",
        "        current_min, current_max = np.nanmin(zscored_df_genera.values), np.nanmax(zscored_df_genera.values)\n",
        "        if not np.isnan(current_min): global_min_max_genera['min'] = min(global_min_max_genera['min'], current_min)\n",
        "        if not np.isnan(current_max): global_min_max_genera['max'] = max(global_min_max_genera['max'], current_max)\n",
        "\n",
        "    # same handling\n",
        "    vmin_genera = global_min_max_genera['min'] if np.isfinite(global_min_max_genera['min']) else -1\n",
        "    vmax_genera = global_min_max_genera['max'] if np.isfinite(global_min_max_genera['max']) else 1\n",
        "    if vmin_genera == vmax_genera: vmin_genera -= 0.5; vmax_genera += 0.5\n",
        "\n",
        "# plotting from here\n",
        "plot_order = ['Faecal', 'Ileum', 'Duodenum']\n",
        "valid_plots_genera = [loc for loc in plot_order if loc in aligned_zscored_data_genera] # again, new dict\n",
        "num_plots_genera = len(valid_plots_genera) # new var...\n",
        "\n",
        "if num_plots_genera == 0:\n",
        "    print(\"Error (Genus): No valid data to plot.\")\n",
        "else:\n",
        "    num_genera = len(sorted_genera) # do not use the same var\n",
        "    num_conditions = 2\n",
        "    cell_width_inches = 0.2 # more genera\n",
        "    max_fig_width = 40 # do not touch this\n",
        "    fig_width_genera = min(max_fig_width, max(8, cell_width_inches * num_genera + 2))\n",
        "    fig_height_genera = cell_width_inches * num_conditions * num_plots_genera + 2.5\n",
        "\n",
        "    print(f\"creating genus heatmap figure with size...: ({fig_width_genera:.2f}, {fig_height_genera:.2f}) inches\")\n",
        "\n",
        "    fig_genera = plt.figure(figsize=(fig_width_genera, fig_height_genera))\n",
        "    gs_genera = gridspec.GridSpec(num_plots_genera, 1, hspace=0.05, top=0.90, bottom=0.35)\n",
        "\n",
        "    cmap = sns.diverging_palette(240, 10, as_cmap=True)\n",
        "    cmap = cmap.copy()\n",
        "    cmap.set_bad(color='lightgrey')\n",
        "\n",
        "    ax1_genera = None\n",
        "\n",
        "    for i, loc_name in enumerate(valid_plots_genera): # new list\n",
        "        ax_genera = fig_genera.add_subplot(gs_genera[i, 0], sharex=ax1_genera)\n",
        "        if i == 0: ax1_genera = ax_genera\n",
        "\n",
        "        # transpose for Genera as columns\n",
        "        data_to_plot_genera = aligned_zscored_data_genera[loc_name].T # new dict\n",
        "\n",
        "        sns.heatmap(data_to_plot_genera, cmap=cmap, center=0,\n",
        "                    vmin=vmin_genera, vmax=vmax_genera, # new vmin/vmax\n",
        "                    annot=False, linewidths=.5, ax=ax_genera, cbar=False)\n",
        "\n",
        "        ax_genera.set_ylabel(loc_name, rotation=0, ha='right', va='center', labelpad=15, fontweight='bold')\n",
        "        ax_genera.set_title('')\n",
        "        plt.setp(ax_genera.get_yticklabels(), rotation=0)\n",
        "\n",
        "        if i == num_plots_genera - 1:\n",
        "            plt.setp(ax_genera.get_xticklabels(), rotation=90, visible=True, fontsize=8)\n",
        "            ax_genera.set_xlabel('Genus')\n",
        "        else:\n",
        "            plt.setp(ax_genera.get_xticklabels(), visible=False)\n",
        "            ax_genera.set_xlabel('')\n",
        "\n",
        "    cbar_ax_genera = fig_genera.add_axes([0.15, gs_genera.top + 0.02, 0.7, 0.02])\n",
        "    sm_genera = cm.ScalarMappable(cmap=cmap, norm=plt.Normalize(vmin=vmin_genera, vmax=vmax_genera))\n",
        "    sm_genera.set_array([])\n",
        "    cbar_genera = fig_genera.colorbar(sm_genera, cax=cbar_ax_genera, orientation='horizontal')\n",
        "    cbar_genera.ax.xaxis.set_ticks_position('top')\n",
        "    cbar_genera.ax.xaxis.set_label_position('top')\n",
        "    cbar_genera.set_label('', fontsize=10)\n",
        "\n",
        "    fig_genera.suptitle('', y=gs_genera.top + 0.08, fontsize=14)\n",
        "\n",
        "    plt.tight_layout(rect=[0, 0, 1, gs_genera.top])\n",
        "\n",
        "    plt.show()\n",
        "    plt.savefig('fig8_genus_change_heatmap.png', bbox_inches='tight', dpi=300)\n",
        "\"\"\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 122
        },
        "id": "IIHkC9FdMrTs",
        "outputId": "588408ab-c005-4e08-95ce-996965db272b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'\\nfrom scipy.stats import zscore\\nimport matplotlib.pyplot as plt\\nimport matplotlib.gridspec as gridspec\\nimport matplotlib.cm as cm\\nimport numpy as np\\nimport pandas as pd\\n\\n# genus copy\\nlocations_data_genera = {\\n    \\'Faecal\\': merged_df_genera_F.copy() if \\'merged_df_genera_F\\' in locals() and merged_df_genera_F is not None else None,\\n    \\'Ileum\\': merged_df_genera_I.copy() if \\'merged_df_genera_I\\' in locals() and merged_df_genera_I is not None else None,\\n    \\'Duodenum\\': merged_df_genera_ST.copy() if \\'merged_df_genera_ST\\' in locals() and merged_df_genera_ST is not None else None\\n}\\n\\nprocessed_data_genera = {}\\nall_genera = set()\\n\\nwt_groups = [\\'WT_10\\', \\'WT_22\\']\\nhet_groups = [\\'HET_10\\', \\'HET_22\\']\\nrequired_groups = wt_groups + het_groups\\n\\nfor loc_name, loc_df_genera in locations_data_genera.items():\\n    if loc_df_genera is None or loc_df_genera.empty:\\n        print(f\"Skipping {loc_name} (Genus): No data.\")\\n        continue\\n\\n    # check\\n    loc_df_genera.loc[:, \\'Value\\'] = pd.to_numeric(loc_df_genera[\\'Value\\'], errors=\\'coerce\\')\\n\\n    mean_abund_genera = loc_df_genera.groupby([\\'genus\\', \\'Group\\'])[\\'Value\\'].mean().unstack(fill_value=0)\\n\\n    # check\\n    if not all(g in mean_abund_genera.columns for g in required_groups):\\n        print(f\"Skipping {loc_name} (Genus): Missing required groups {required_groups}. Found: {mean_abund_genera.columns.tolist()}\")\\n        continue\\n\\n    # log2 transform. pseudocount 1\\n    log2_abund_genera = mean_abund_genera.apply(lambda x: np.log2(x + 1))\\n\\n    if all(g in log2_abund_genera.columns for g in wt_groups) and all(g in log2_abund_genera.columns for g in het_groups):\\n        delta_df_genera = pd.DataFrame({\\n            \\'WT_change\\': log2_abund_genera[\\'WT_22\\'] - log2_abund_genera[\\'WT_10\\'],\\n            \\'HET_change\\': log2_abund_genera[\\'HET_22\\'] - log2_abund_genera[\\'HET_10\\']\\n        }, index=log2_abund_genera.index) # genus names\\n\\n        processed_data_genera[loc_name] = delta_df_genera\\n        all_genera.update(delta_df_genera.index)\\n    else:\\n         print(f\"Skipping delta calculation for {loc_name} (Genus): Missing one or more required time point groups in log2 data.\")\\n\\n\\n# aligning genera. again check no overwriting BEFORE running please please please.\\naligned_zscored_data_genera = {}\\nglobal_min_max_genera = {\\'min\\': float(\\'inf\\'), \\'max\\': float(\\'-inf\\')}\\n\\nif not all_genera:\\n    print(\"no genera found after processing\")\\nelse:\\n    sorted_genera = sorted(list(all_genera))\\n    print(f\"processing {len(sorted_genera)} genera..\") # check again\\n\\n    for loc_name, delta_df_genera in processed_data_genera.items():\\n        aligned_df_genera = delta_df_genera.reindex(sorted_genera, fill_value=np.nan)\\n\\n        zscored_df_genera = aligned_df_genera.astype(float).apply(zscore, axis=0, nan_policy=\\'omit\\')\\n        aligned_zscored_data_genera[loc_name] = zscored_df_genera\\n\\n        current_min, current_max = np.nanmin(zscored_df_genera.values), np.nanmax(zscored_df_genera.values)\\n        if not np.isnan(current_min): global_min_max_genera[\\'min\\'] = min(global_min_max_genera[\\'min\\'], current_min)\\n        if not np.isnan(current_max): global_min_max_genera[\\'max\\'] = max(global_min_max_genera[\\'max\\'], current_max)\\n\\n    # same handling\\n    vmin_genera = global_min_max_genera[\\'min\\'] if np.isfinite(global_min_max_genera[\\'min\\']) else -1\\n    vmax_genera = global_min_max_genera[\\'max\\'] if np.isfinite(global_min_max_genera[\\'max\\']) else 1\\n    if vmin_genera == vmax_genera: vmin_genera -= 0.5; vmax_genera += 0.5\\n\\n# plotting from here\\nplot_order = [\\'Faecal\\', \\'Ileum\\', \\'Duodenum\\']\\nvalid_plots_genera = [loc for loc in plot_order if loc in aligned_zscored_data_genera] # again, new dict\\nnum_plots_genera = len(valid_plots_genera) # new var...\\n\\nif num_plots_genera == 0:\\n    print(\"Error (Genus): No valid data to plot.\")\\nelse:\\n    num_genera = len(sorted_genera) # do not use the same var\\n    num_conditions = 2\\n    cell_width_inches = 0.2 # more genera\\n    max_fig_width = 40 # do not touch this\\n    fig_width_genera = min(max_fig_width, max(8, cell_width_inches * num_genera + 2))\\n    fig_height_genera = cell_width_inches * num_conditions * num_plots_genera + 2.5\\n\\n    print(f\"creating genus heatmap figure with size...: ({fig_width_genera:.2f}, {fig_height_genera:.2f}) inches\")\\n\\n    fig_genera = plt.figure(figsize=(fig_width_genera, fig_height_genera))\\n    gs_genera = gridspec.GridSpec(num_plots_genera, 1, hspace=0.05, top=0.90, bottom=0.35)\\n\\n    cmap = sns.diverging_palette(240, 10, as_cmap=True)\\n    cmap = cmap.copy()\\n    cmap.set_bad(color=\\'lightgrey\\')\\n\\n    ax1_genera = None\\n\\n    for i, loc_name in enumerate(valid_plots_genera): # new list\\n        ax_genera = fig_genera.add_subplot(gs_genera[i, 0], sharex=ax1_genera)\\n        if i == 0: ax1_genera = ax_genera\\n\\n        # transpose for Genera as columns\\n        data_to_plot_genera = aligned_zscored_data_genera[loc_name].T # new dict\\n\\n        sns.heatmap(data_to_plot_genera, cmap=cmap, center=0,\\n                    vmin=vmin_genera, vmax=vmax_genera, # new vmin/vmax\\n                    annot=False, linewidths=.5, ax=ax_genera, cbar=False)\\n        \\n        ax_genera.set_ylabel(loc_name, rotation=0, ha=\\'right\\', va=\\'center\\', labelpad=15, fontweight=\\'bold\\')\\n        ax_genera.set_title(\\'\\')\\n        plt.setp(ax_genera.get_yticklabels(), rotation=0)\\n\\n        if i == num_plots_genera - 1:\\n            plt.setp(ax_genera.get_xticklabels(), rotation=90, visible=True, fontsize=8)\\n            ax_genera.set_xlabel(\\'Genus\\')\\n        else:\\n            plt.setp(ax_genera.get_xticklabels(), visible=False)\\n            ax_genera.set_xlabel(\\'\\')\\n\\n    cbar_ax_genera = fig_genera.add_axes([0.15, gs_genera.top + 0.02, 0.7, 0.02])\\n    sm_genera = cm.ScalarMappable(cmap=cmap, norm=plt.Normalize(vmin=vmin_genera, vmax=vmax_genera))\\n    sm_genera.set_array([])\\n    cbar_genera = fig_genera.colorbar(sm_genera, cax=cbar_ax_genera, orientation=\\'horizontal\\')\\n    cbar_genera.ax.xaxis.set_ticks_position(\\'top\\')\\n    cbar_genera.ax.xaxis.set_label_position(\\'top\\')\\n    cbar_genera.set_label(\\'\\', fontsize=10)\\n\\n    fig_genera.suptitle(\\'\\', y=gs_genera.top + 0.08, fontsize=14)\\n\\n    plt.tight_layout(rect=[0, 0, 1, gs_genera.top])\\n\\n    plt.show()\\n    plt.savefig(\\'fig8_genus_change_heatmap.png\\', bbox_inches=\\'tight\\', dpi=300)\\n'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 177
        }
      ]
    }
  ]
}